// @optique/core/dist/constructs.cjs
const require_message = require('./message.cjs');
const require_dependency = require('./dependency.cjs');
const require_mode_dispatch = require('./mode-dispatch.cjs');
const require_usage = require('./usage.cjs');
const require_suggestion = require('./suggestion.cjs');
const require_usage_internals = require('./usage-internals.cjs');

//#region src/constructs.ts
function createUnexpectedInputErrorWithScopedSuggestions(baseError, invalidInput, parsers, customFormatter) {
	const options = /* @__PURE__ */ new Set();
	const commands = /* @__PURE__ */ new Set();
	for (const parser of parsers) require_usage_internals.collectLeadingCandidates(parser.usage, options, commands);
	const candidates = new Set([...options, ...commands]);
	const suggestions = require_suggestion.findSimilar(invalidInput, candidates, require_suggestion.DEFAULT_FIND_SIMILAR_OPTIONS);
	const suggestionMsg = customFormatter ? customFormatter(suggestions) : require_suggestion.createSuggestionMessage(suggestions);
	return suggestionMsg.length > 0 ? [
		...baseError,
		require_message.text("\n\n"),
		...suggestionMsg
	] : baseError;
}
/**
* Checks if the given token is an option name that requires a value
* (i.e., has a metavar) within the given usage terms.
* @param usage The usage terms to search through.
* @param token The token to check.
* @returns `true` if the token is an option that requires a value, `false` otherwise.
*/
function isOptionRequiringValue(usage, token) {
	function traverse(terms) {
		if (!terms || !Array.isArray(terms)) return false;
		for (const term of terms) if (term.type === "option") {
			if (term.metavar && term.names.includes(token)) return true;
		} else if (term.type === "optional" || term.type === "multiple") {
			if (traverse(term.terms)) return true;
		} else if (term.type === "exclusive") {
			for (const exclusiveUsage of term.terms) if (traverse(exclusiveUsage)) return true;
		}
		return false;
	}
	return traverse(usage);
}
/**
* Extracts required (non-optional) usage terms from a usage array.
* @param usage The usage to extract required terms from
* @returns Usage containing only required (non-optional) terms
*/
function extractRequiredUsage(usage) {
	const required = [];
	for (const term of usage) if (term.type === "optional") continue;
	else if (term.type === "exclusive") {
		const requiredBranches = term.terms.map((branch) => extractRequiredUsage(branch)).filter((branch) => branch.length > 0);
		if (requiredBranches.length > 0) required.push({
			type: "exclusive",
			terms: requiredBranches
		});
	} else if (term.type === "multiple") {
		if (term.min > 0) {
			const requiredTerms = extractRequiredUsage(term.terms);
			if (requiredTerms.length > 0) required.push({
				type: "multiple",
				terms: requiredTerms,
				min: term.min
			});
		}
	} else required.push(term);
	return required;
}
/**
* Analyzes parsers to determine what types of inputs are expected.
* @param parsers The parsers being combined
* @returns Context about what types of inputs are expected
*/
function analyzeNoMatchContext(parsers) {
	const combinedUsage = [{
		type: "exclusive",
		terms: parsers.map((p) => p.usage)
	}];
	const requiredUsage = extractRequiredUsage(combinedUsage);
	return {
		hasOptions: require_usage.extractOptionNames(requiredUsage).size > 0,
		hasCommands: require_usage.extractCommandNames(requiredUsage).size > 0,
		hasArguments: require_usage.extractArgumentMetavars(requiredUsage).size > 0
	};
}
/**
* Error class thrown when duplicate option names are detected during parser
* construction. This is a programmer error, not a user error.
*/
var DuplicateOptionError = class extends Error {
	constructor(optionName$1, sources) {
		const sourceNames = sources.map((s) => typeof s === "symbol" ? s.description ?? s.toString() : s);
		super(`Duplicate option name "${optionName$1}" found in fields: ${sourceNames.join(", ")}. Each option name must be unique within a parser combinator.`);
		this.optionName = optionName$1;
		this.sources = sources;
		this.name = "DuplicateOptionError";
	}
};
/**
* Checks for duplicate option names across parser sources and throws an error
* if duplicates are found. This should be called at construction time.
* @param parserSources Array of [source, usage] tuples
* @throws DuplicateOptionError if duplicate option names are found
*/
function checkDuplicateOptionNames(parserSources) {
	const optionNameSources = /* @__PURE__ */ new Map();
	for (const [source, usage] of parserSources) {
		const names = require_usage.extractOptionNames(usage);
		for (const name of names) {
			if (!optionNameSources.has(name)) optionNameSources.set(name, []);
			optionNameSources.get(name).push(source);
		}
	}
	for (const [name, sources] of optionNameSources) if (sources.length > 1) throw new DuplicateOptionError(name, sources);
}
/**
* Generates a contextual error message based on what types of inputs
* the parsers expect (options, commands, or arguments).
* @param context Context about what types of inputs are expected
* @returns An appropriate error message
*/
function generateNoMatchError(context) {
	const { hasOptions, hasCommands, hasArguments } = context;
	if (hasArguments && !hasOptions && !hasCommands) return require_message.message`Missing required argument.`;
	else if (hasCommands && !hasOptions && !hasArguments) return require_message.message`No matching command found.`;
	else if (hasOptions && !hasCommands && !hasArguments) return require_message.message`No matching option found.`;
	else if (hasCommands && hasOptions && !hasArguments) return require_message.message`No matching option or command found.`;
	else if (hasArguments && hasOptions && !hasCommands) return require_message.message`No matching option or argument found.`;
	else if (hasArguments && hasCommands && !hasOptions) return require_message.message`No matching command or argument found.`;
	else return require_message.message`No matching option, command, or argument found.`;
}
/**
* Creates a complete() method shared by or() and longestMatch().
* @internal
*/
function createExclusiveComplete(parsers, options, noMatchContext, mode) {
	const syncParsers = parsers;
	return (state) => {
		if (state == null) return {
			success: false,
			error: getNoMatchError(options, noMatchContext)
		};
		const [i, result] = state;
		if (!result.success) return {
			success: false,
			error: result.error
		};
		return require_mode_dispatch.dispatchByMode(mode, () => syncParsers[i].complete(result.next.state), async () => {
			const completeResult = await parsers[i].complete(result.next.state);
			return completeResult;
		});
	};
}
/**
* Creates a suggest() method shared by or() and longestMatch().
* @internal
*/
function createExclusiveSuggest(parsers, mode) {
	const syncParsers = parsers;
	return (context, prefix) => {
		return require_mode_dispatch.dispatchIterableByMode(mode, function* () {
			const suggestions = [];
			if (context.state == null) for (const parser of syncParsers) {
				const parserSuggestions = parser.suggest({
					...context,
					state: parser.initialState
				}, prefix);
				suggestions.push(...parserSuggestions);
			}
			else {
				const [index, parserResult] = context.state;
				if (parserResult.success) {
					const parserSuggestions = syncParsers[index].suggest({
						...context,
						state: parserResult.next.state
					}, prefix);
					suggestions.push(...parserSuggestions);
				}
			}
			yield* require_suggestion.deduplicateSuggestions(suggestions);
		}, async function* () {
			const suggestions = [];
			if (context.state == null) for (const parser of parsers) {
				const parserSuggestions = parser.suggest({
					...context,
					state: parser.initialState
				}, prefix);
				if (parser.$mode === "async") for await (const s of parserSuggestions) suggestions.push(s);
				else suggestions.push(...parserSuggestions);
			}
			else {
				const [index, parserResult] = context.state;
				if (parserResult.success) {
					const parser = parsers[index];
					const parserSuggestions = parser.suggest({
						...context,
						state: parserResult.next.state
					}, prefix);
					if (parser.$mode === "async") for await (const s of parserSuggestions) suggestions.push(s);
					else suggestions.push(...parserSuggestions);
				}
			}
			yield* require_suggestion.deduplicateSuggestions(suggestions);
		});
	};
}
/**
* Gets the no-match error, either from custom options or default.
* Shared by or() and longestMatch().
* @internal
*/
function getNoMatchError(options, noMatchContext) {
	const customNoMatch = options?.errors?.noMatch;
	return customNoMatch ? typeof customNoMatch === "function" ? customNoMatch(noMatchContext) : customNoMatch : generateNoMatchError(noMatchContext);
}
/**
* @since 0.5.0
*/
function or(...args) {
	let parsers;
	let options;
	if (args.length > 0 && args[args.length - 1] && typeof args[args.length - 1] === "object" && !("$valueType" in args[args.length - 1])) {
		options = args[args.length - 1];
		parsers = args.slice(0, -1);
	} else {
		parsers = args;
		options = void 0;
	}
	const noMatchContext = analyzeNoMatchContext(parsers);
	const combinedMode = parsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const syncParsers = parsers;
	const getInitialError = (context) => ({
		consumed: 0,
		error: context.buffer.length < 1 ? getNoMatchError(options, noMatchContext) : (() => {
			const token = context.buffer[0];
			const defaultMsg = require_message.message`Unexpected option or subcommand: ${require_message.optionName(token)}.`;
			if (options?.errors?.unexpectedInput != null) return typeof options.errors.unexpectedInput === "function" ? options.errors.unexpectedInput(token) : options.errors.unexpectedInput;
			return createUnexpectedInputErrorWithScopedSuggestions(defaultMsg, token, parsers, options?.errors?.suggestions);
		})()
	});
	const parseSync = (context) => {
		let error = getInitialError(context);
		const orderedParsers = syncParsers.map((p, i) => [p, i]);
		orderedParsers.sort(([_, a], [__, b]) => context.state?.[0] === a ? -1 : context.state?.[0] === b ? 1 : a - b);
		for (const [parser, i] of orderedParsers) {
			const result = parser.parse({
				...context,
				state: context.state == null || context.state[0] !== i || !context.state[1].success ? parser.initialState : context.state[1].next.state
			});
			if (result.success && result.consumed.length > 0) {
				if (context.state?.[0] !== i && context.state?.[1].success) {
					const previouslyConsumed = context.state[1].consumed;
					const checkResult = parser.parse({
						...context,
						buffer: previouslyConsumed,
						state: parser.initialState
					});
					const canConsumeShared = checkResult.success && checkResult.consumed.length === previouslyConsumed.length && checkResult.consumed.every((c, idx) => c === previouslyConsumed[idx]);
					if (!canConsumeShared) return {
						success: false,
						consumed: context.buffer.length - result.next.buffer.length,
						error: require_message.message`${require_message.values(context.state[1].consumed)} and ${require_message.values(result.consumed)} cannot be used together.`
					};
					const replayedResult = parser.parse({
						...context,
						state: checkResult.next.state
					});
					if (!replayedResult.success) return replayedResult;
					return {
						success: true,
						next: {
							...context,
							buffer: replayedResult.next.buffer,
							optionsTerminated: replayedResult.next.optionsTerminated,
							state: [i, {
								...replayedResult,
								consumed: [...previouslyConsumed, ...replayedResult.consumed]
							}]
						},
						consumed: replayedResult.consumed
					};
				}
				return {
					success: true,
					next: {
						...context,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: [i, result]
					},
					consumed: result.consumed
				};
			} else if (!result.success && error.consumed < result.consumed) error = result;
		}
		return {
			...error,
			success: false
		};
	};
	const parseAsync = async (context) => {
		let error = getInitialError(context);
		const orderedParsers = parsers.map((p, i) => [p, i]);
		orderedParsers.sort(([_, a], [__, b]) => context.state?.[0] === a ? -1 : context.state?.[0] === b ? 1 : a - b);
		for (const [parser, i] of orderedParsers) {
			const resultOrPromise = parser.parse({
				...context,
				state: context.state == null || context.state[0] !== i || !context.state[1].success ? parser.initialState : context.state[1].next.state
			});
			const result = await resultOrPromise;
			if (result.success && result.consumed.length > 0) {
				if (context.state?.[0] !== i && context.state?.[1].success) {
					const previouslyConsumed = context.state[1].consumed;
					const checkResultOrPromise = parser.parse({
						...context,
						buffer: previouslyConsumed,
						state: parser.initialState
					});
					const checkResult = await checkResultOrPromise;
					const canConsumeShared = checkResult.success && checkResult.consumed.length === previouslyConsumed.length && checkResult.consumed.every((c, idx) => c === previouslyConsumed[idx]);
					if (!canConsumeShared) return {
						success: false,
						consumed: context.buffer.length - result.next.buffer.length,
						error: require_message.message`${require_message.values(context.state[1].consumed)} and ${require_message.values(result.consumed)} cannot be used together.`
					};
					const replayedResultOrPromise = parser.parse({
						...context,
						state: checkResult.next.state
					});
					const replayedResult = await replayedResultOrPromise;
					if (!replayedResult.success) return replayedResult;
					return {
						success: true,
						next: {
							...context,
							buffer: replayedResult.next.buffer,
							optionsTerminated: replayedResult.next.optionsTerminated,
							state: [i, {
								...replayedResult,
								consumed: [...previouslyConsumed, ...replayedResult.consumed]
							}]
						},
						consumed: replayedResult.consumed
					};
				}
				return {
					success: true,
					next: {
						...context,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: [i, result]
					},
					consumed: result.consumed
				};
			} else if (!result.success && error.consumed < result.consumed) error = result;
		}
		return {
			...error,
			success: false
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: Math.max(...parsers.map((p) => p.priority)),
		usage: [{
			type: "exclusive",
			terms: parsers.map((p) => p.usage)
		}],
		initialState: void 0,
		complete: createExclusiveComplete(parsers, options, noMatchContext, combinedMode),
		parse(context) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => parseSync(context), () => parseAsync(context));
		},
		suggest: createExclusiveSuggest(parsers, combinedMode),
		getDocFragments(state, _defaultValue) {
			let brief;
			let description;
			let footer;
			let fragments;
			if (state.kind === "unavailable" || state.state == null) fragments = parsers.flatMap((p) => p.getDocFragments({ kind: "unavailable" }, void 0).fragments);
			else {
				const [index, parserResult] = state.state;
				const innerState = parserResult.success ? {
					kind: "available",
					state: parserResult.next.state
				} : { kind: "unavailable" };
				const docFragments = parsers[index].getDocFragments(innerState, void 0);
				brief = docFragments.brief;
				description = docFragments.description;
				footer = docFragments.footer;
				fragments = docFragments.fragments;
			}
			const entries = fragments.filter((f) => f.type === "entry");
			const sections = [];
			for (const fragment of fragments) {
				if (fragment.type !== "section") continue;
				if (fragment.title == null) entries.push(...fragment.entries);
				else sections.push(fragment);
			}
			return {
				brief,
				description,
				footer,
				fragments: [...sections.map((s) => ({
					...s,
					type: "section"
				})), {
					type: "section",
					entries
				}]
			};
		}
	};
}
/**
* @since 0.5.0
*/
function longestMatch(...args) {
	let parsers;
	let options;
	if (args.length > 0 && args[args.length - 1] && typeof args[args.length - 1] === "object" && !("$valueType" in args[args.length - 1])) {
		options = args[args.length - 1];
		parsers = args.slice(0, -1);
	} else {
		parsers = args;
		options = void 0;
	}
	const noMatchContext = analyzeNoMatchContext(parsers);
	const combinedMode = parsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const syncParsers = parsers;
	const getInitialError = (context) => ({
		consumed: 0,
		error: context.buffer.length < 1 ? getNoMatchError(options, noMatchContext) : (() => {
			const token = context.buffer[0];
			const defaultMsg = require_message.message`Unexpected option or subcommand: ${require_message.optionName(token)}.`;
			if (options?.errors?.unexpectedInput != null) return typeof options.errors.unexpectedInput === "function" ? options.errors.unexpectedInput(token) : options.errors.unexpectedInput;
			return createUnexpectedInputErrorWithScopedSuggestions(defaultMsg, token, parsers, options?.errors?.suggestions);
		})()
	});
	const parseSync = (context) => {
		let bestMatch = null;
		let error = getInitialError(context);
		for (let i = 0; i < syncParsers.length; i++) {
			const parser = syncParsers[i];
			const result = parser.parse({
				...context,
				state: context.state == null || context.state[0] !== i || !context.state[1].success ? parser.initialState : context.state[1].next.state
			});
			if (result.success) {
				const consumed = context.buffer.length - result.next.buffer.length;
				if (bestMatch === null || consumed > bestMatch.consumed) bestMatch = {
					index: i,
					result,
					consumed
				};
			} else if (error.consumed < result.consumed) error = result;
		}
		if (bestMatch && bestMatch.result.success) return {
			success: true,
			next: {
				...context,
				buffer: bestMatch.result.next.buffer,
				optionsTerminated: bestMatch.result.next.optionsTerminated,
				state: [bestMatch.index, bestMatch.result]
			},
			consumed: bestMatch.result.consumed
		};
		return {
			...error,
			success: false
		};
	};
	const parseAsync = async (context) => {
		let bestMatch = null;
		let error = getInitialError(context);
		for (let i = 0; i < parsers.length; i++) {
			const parser = parsers[i];
			const resultOrPromise = parser.parse({
				...context,
				state: context.state == null || context.state[0] !== i || !context.state[1].success ? parser.initialState : context.state[1].next.state
			});
			const result = await resultOrPromise;
			if (result.success) {
				const consumed = context.buffer.length - result.next.buffer.length;
				if (bestMatch === null || consumed > bestMatch.consumed) bestMatch = {
					index: i,
					result,
					consumed
				};
			} else if (error.consumed < result.consumed) error = result;
		}
		if (bestMatch && bestMatch.result.success) return {
			success: true,
			next: {
				...context,
				buffer: bestMatch.result.next.buffer,
				optionsTerminated: bestMatch.result.next.optionsTerminated,
				state: [bestMatch.index, bestMatch.result]
			},
			consumed: bestMatch.result.consumed
		};
		return {
			...error,
			success: false
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: Math.max(...parsers.map((p) => p.priority)),
		usage: [{
			type: "exclusive",
			terms: parsers.map((p) => p.usage)
		}],
		initialState: void 0,
		complete: createExclusiveComplete(parsers, options, noMatchContext, combinedMode),
		parse(context) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => parseSync(context), () => parseAsync(context));
		},
		suggest: createExclusiveSuggest(parsers, combinedMode),
		getDocFragments(state, _defaultValue) {
			let brief;
			let description;
			let footer;
			let fragments;
			if (state.kind === "unavailable" || state.state == null) fragments = parsers.flatMap((p) => p.getDocFragments({ kind: "unavailable" }).fragments);
			else {
				const [i, result] = state.state;
				if (result.success) {
					const docResult = parsers[i].getDocFragments({
						kind: "available",
						state: result.next.state
					});
					brief = docResult.brief;
					description = docResult.description;
					footer = docResult.footer;
					fragments = docResult.fragments;
				} else fragments = parsers.flatMap((p) => p.getDocFragments({ kind: "unavailable" }).fragments);
			}
			return {
				brief,
				description,
				fragments,
				footer
			};
		}
	};
}
/**
* Internal sync helper for object suggest functionality.
* @internal
*/
function* suggestObjectSync(context, prefix, parserPairs) {
	const registry = context.dependencyRegistry instanceof require_dependency.DependencyRegistry ? context.dependencyRegistry : new require_dependency.DependencyRegistry();
	if (context.state && typeof context.state === "object") collectDependencies(context.state, registry);
	const contextWithRegistry = {
		...context,
		dependencyRegistry: registry
	};
	if (context.buffer.length > 0) {
		const lastToken = context.buffer[context.buffer.length - 1];
		for (const [field, parser] of parserPairs) if (isOptionRequiringValue(parser.usage, lastToken)) {
			const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
			yield* parser.suggest({
				...contextWithRegistry,
				state: fieldState
			}, prefix);
			return;
		}
	}
	const suggestions = [];
	for (const [field, parser] of parserPairs) {
		const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
		const fieldSuggestions = parser.suggest({
			...contextWithRegistry,
			state: fieldState
		}, prefix);
		suggestions.push(...fieldSuggestions);
	}
	yield* require_suggestion.deduplicateSuggestions(suggestions);
}
/**
* Internal async helper for object suggest functionality.
* @internal
*/
async function* suggestObjectAsync(context, prefix, parserPairs) {
	const registry = context.dependencyRegistry instanceof require_dependency.DependencyRegistry ? context.dependencyRegistry : new require_dependency.DependencyRegistry();
	if (context.state && typeof context.state === "object") collectDependencies(context.state, registry);
	const contextWithRegistry = {
		...context,
		dependencyRegistry: registry
	};
	if (context.buffer.length > 0) {
		const lastToken = context.buffer[context.buffer.length - 1];
		for (const [field, parser] of parserPairs) if (isOptionRequiringValue(parser.usage, lastToken)) {
			const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
			const suggestions$1 = parser.suggest({
				...contextWithRegistry,
				state: fieldState
			}, prefix);
			for await (const s of suggestions$1) yield s;
			return;
		}
	}
	const suggestions = [];
	for (const [field, parser] of parserPairs) {
		const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
		const fieldSuggestions = parser.suggest({
			...contextWithRegistry,
			state: fieldState
		}, prefix);
		for await (const s of fieldSuggestions) suggestions.push(s);
	}
	yield* require_suggestion.deduplicateSuggestions(suggestions);
}
/**
* Resolves deferred parse states in an object's field states.
* This function builds a dependency registry from DependencySourceState fields
* and re-parses DeferredParseState fields using the actual dependency values.
*
* @param fieldStates A record of field names to their state values
* @returns The field states with deferred states resolved to their actual values
* @internal
*/
/**
* Recursively collects dependency values from DependencySourceState objects
* found anywhere in the state tree.
*/
function collectDependencies(state, registry, visited = /* @__PURE__ */ new WeakSet()) {
	if (state === null || state === void 0) return;
	if (typeof state === "object") {
		if (visited.has(state)) return;
		visited.add(state);
	}
	if (require_dependency.isDependencySourceState(state)) {
		const depId = state[require_dependency.dependencyId];
		const result = state.result;
		if (result.success) registry.set(depId, result.value);
		return;
	}
	if (Array.isArray(state)) {
		for (const item of state) collectDependencies(item, registry, visited);
		return;
	}
	if (typeof state === "object" && !require_dependency.isDeferredParseState(state)) for (const key of Reflect.ownKeys(state)) collectDependencies(state[key], registry, visited);
}
/**
* Checks if a value is a plain object (created with `{}` or `Object.create(null)`).
* Class instances like `Temporal.PlainDate`, `URL`, `Date`, etc. return false.
* This is used to determine whether to recursively traverse an object when
* resolving deferred parse states - we only want to traverse plain objects
* that are part of the parser state structure, not user values.
*/
function isPlainObject(value) {
	if (typeof value !== "object" || value === null) return false;
	const proto = Object.getPrototypeOf(value);
	return proto === Object.prototype || proto === null;
}
/**
* Collects dependency values for a DeferredParseState from the registry.
* Returns the collected values array, or null if any required dependency
* is missing (and no default is available).
*/
function collectDependencyValues(deferredState, registry) {
	const depIds = deferredState.dependencyIds;
	if (depIds && depIds.length > 0) {
		const defaults = deferredState.defaultValues;
		const dependencyValues = [];
		for (let i = 0; i < depIds.length; i++) {
			const depId$1 = depIds[i];
			if (registry.has(depId$1)) dependencyValues.push(registry.get(depId$1));
			else if (defaults && i < defaults.length) dependencyValues.push(defaults[i]);
			else return null;
		}
		return dependencyValues;
	}
	const depId = deferredState.dependencyId;
	if (registry.has(depId)) return registry.get(depId);
	return null;
}
/**
* Recursively resolves DeferredParseState objects found anywhere in the state tree.
* Returns the resolved state (sync version).
*
* Only traverses:
* - DeferredParseState (to resolve it)
* - DependencySourceState (skipped, kept as-is)
* - Arrays (to find nested deferred states)
* - Plain objects (to find nested deferred states in parser state structures)
*
* Does NOT traverse class instances (e.g., Temporal.PlainDate, URL) since these
* are user values that should be preserved as-is.
*/
function resolveDeferred(state, registry, visited = /* @__PURE__ */ new WeakSet()) {
	if (state === null || state === void 0) return state;
	if (typeof state === "object") {
		if (visited.has(state)) return state;
		visited.add(state);
	}
	if (require_dependency.isDeferredParseState(state)) {
		const deferredState = state;
		const dependencyValue = collectDependencyValues(deferredState, registry);
		if (dependencyValue === null) return deferredState.preliminaryResult;
		const reParseResult = deferredState.parser[require_dependency.parseWithDependency](deferredState.rawInput, dependencyValue);
		if (reParseResult instanceof Promise) return deferredState.preliminaryResult;
		return reParseResult;
	}
	if (require_dependency.isDependencySourceState(state)) return state;
	if (Array.isArray(state)) return state.map((item) => resolveDeferred(item, registry, visited));
	if (isPlainObject(state)) {
		const resolved = {};
		for (const key of Reflect.ownKeys(state)) resolved[key] = resolveDeferred(state[key], registry, visited);
		return resolved;
	}
	return state;
}
function resolveDeferredParseStates(fieldStates) {
	const registry = new require_dependency.DependencyRegistry();
	collectDependencies(fieldStates, registry);
	return resolveDeferred(fieldStates, registry);
}
/**
* Recursively resolves DeferredParseState objects found anywhere in the state tree.
* Returns the resolved state (async version).
*
* Only traverses:
* - DeferredParseState (to resolve it)
* - DependencySourceState (skipped, kept as-is)
* - Arrays (to find nested deferred states)
* - Plain objects (to find nested deferred states in parser state structures)
*
* Does NOT traverse class instances (e.g., Temporal.PlainDate, URL) since these
* are user values that should be preserved as-is.
*/
async function resolveDeferredAsync(state, registry, visited = /* @__PURE__ */ new WeakSet()) {
	if (state === null || state === void 0) return state;
	if (typeof state === "object") {
		if (visited.has(state)) return state;
		visited.add(state);
	}
	if (require_dependency.isDeferredParseState(state)) {
		const deferredState = state;
		const dependencyValue = collectDependencyValues(deferredState, registry);
		if (dependencyValue === null) return deferredState.preliminaryResult;
		const reParseResult = deferredState.parser[require_dependency.parseWithDependency](deferredState.rawInput, dependencyValue);
		return Promise.resolve(reParseResult);
	}
	if (require_dependency.isDependencySourceState(state)) return state;
	if (Array.isArray(state)) return Promise.all(state.map((item) => resolveDeferredAsync(item, registry, visited)));
	if (isPlainObject(state)) {
		const resolved = {};
		const keys = Reflect.ownKeys(state);
		await Promise.all(keys.map(async (key) => {
			resolved[key] = await resolveDeferredAsync(state[key], registry, visited);
		}));
		return resolved;
	}
	return state;
}
/**
* Async version of resolveDeferredParseStates for async parsers.
* @internal
*/
async function resolveDeferredParseStatesAsync(fieldStates) {
	const registry = new require_dependency.DependencyRegistry();
	collectDependencies(fieldStates, registry);
	return await resolveDeferredAsync(fieldStates, registry);
}
function object(labelOrParsers, maybeParsersOrOptions, maybeOptions) {
	const label = typeof labelOrParsers === "string" ? labelOrParsers : void 0;
	let parsers;
	let options = {};
	if (typeof labelOrParsers === "string") {
		parsers = maybeParsersOrOptions;
		options = maybeOptions ?? {};
	} else {
		parsers = labelOrParsers;
		options = maybeParsersOrOptions ?? {};
	}
	const parserKeys = Reflect.ownKeys(parsers);
	const parserPairs = parserKeys.map((k) => [k, parsers[k]]);
	parserPairs.sort(([_, parserA], [__, parserB]) => parserB.priority - parserA.priority);
	const initialState = {};
	for (const key of parserKeys) initialState[key] = parsers[key].initialState;
	if (!options.allowDuplicates) checkDuplicateOptionNames(parserPairs.map(([field, parser]) => [field, parser.usage]));
	const noMatchContext = analyzeNoMatchContext(parserKeys.map((k) => parsers[k]));
	const combinedMode = parserKeys.some((k) => parsers[k].$mode === "async") ? "async" : "sync";
	const getInitialError = (context) => ({
		consumed: 0,
		error: context.buffer.length > 0 ? (() => {
			const token = context.buffer[0];
			const customMessage = options.errors?.unexpectedInput;
			if (customMessage) return typeof customMessage === "function" ? customMessage(token) : customMessage;
			const baseError = require_message.message`Unexpected option or argument: ${token}.`;
			return require_suggestion.createErrorWithSuggestions(baseError, token, context.usage, "both", options.errors?.suggestions);
		})() : (() => {
			const customEndOfInput = options.errors?.endOfInput;
			return customEndOfInput ? typeof customEndOfInput === "function" ? customEndOfInput(noMatchContext) : customEndOfInput : generateNoMatchError(noMatchContext);
		})()
	});
	const parseSync = (context) => {
		let error = getInitialError(context);
		let currentContext = context;
		let anySuccess = false;
		const allConsumed = [];
		let madeProgress = true;
		while (madeProgress && currentContext.buffer.length > 0) {
			madeProgress = false;
			for (const [field, parser] of parserPairs) {
				const result = parser.parse({
					...currentContext,
					state: currentContext.state && typeof currentContext.state === "object" && field in currentContext.state ? currentContext.state[field] : parser.initialState
				});
				if (result.success && result.consumed.length > 0) {
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: {
							...currentContext.state,
							[field]: result.next.state
						}
					};
					allConsumed.push(...result.consumed);
					anySuccess = true;
					madeProgress = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
		}
		if (anySuccess) return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
		if (context.buffer.length === 0) {
			let allCanComplete = true;
			for (const [field, parser] of parserPairs) {
				const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
				const completeResult = parser.complete(fieldState);
				if (!completeResult.success) {
					allCanComplete = false;
					break;
				}
			}
			if (allCanComplete) return {
				success: true,
				next: context,
				consumed: []
			};
		}
		return {
			...error,
			success: false
		};
	};
	const parseAsync = async (context) => {
		let error = getInitialError(context);
		let currentContext = context;
		let anySuccess = false;
		const allConsumed = [];
		let madeProgress = true;
		while (madeProgress && currentContext.buffer.length > 0) {
			madeProgress = false;
			for (const [field, parser] of parserPairs) {
				const resultOrPromise = parser.parse({
					...currentContext,
					state: currentContext.state && typeof currentContext.state === "object" && field in currentContext.state ? currentContext.state[field] : parser.initialState
				});
				const result = await resultOrPromise;
				if (result.success && result.consumed.length > 0) {
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: {
							...currentContext.state,
							[field]: result.next.state
						}
					};
					allConsumed.push(...result.consumed);
					anySuccess = true;
					madeProgress = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
		}
		if (anySuccess) return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
		if (context.buffer.length === 0) {
			let allCanComplete = true;
			for (const [field, parser] of parserPairs) {
				const fieldState = context.state && typeof context.state === "object" && field in context.state ? context.state[field] : parser.initialState;
				const completeResult = await parser.complete(fieldState);
				if (!completeResult.success) {
					allCanComplete = false;
					break;
				}
			}
			if (allCanComplete) return {
				success: true,
				next: context,
				consumed: []
			};
		}
		return {
			...error,
			success: false
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: Math.max(...parserKeys.map((k) => parsers[k].priority)),
		usage: parserPairs.flatMap(([_, p]) => p.usage),
		initialState,
		parse(context) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => parseSync(context), () => parseAsync(context));
		},
		complete(state) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => {
				const preCompletedState = {};
				const preCompletedKeys = /* @__PURE__ */ new Set();
				for (const field of parserKeys) {
					const fieldKey = field;
					const fieldState = state[fieldKey];
					const fieldParser = parsers[field];
					if (Array.isArray(fieldState) && fieldState.length === 1 && require_dependency.isPendingDependencySourceState(fieldState[0])) {
						const completed = fieldParser.complete(fieldState);
						preCompletedState[fieldKey] = completed;
						preCompletedKeys.add(fieldKey);
					} else if (fieldState === void 0 && require_dependency.isPendingDependencySourceState(fieldParser.initialState)) {
						const completed = fieldParser.complete([fieldParser.initialState]);
						preCompletedState[fieldKey] = completed;
						preCompletedKeys.add(fieldKey);
					} else if (fieldState === void 0 && require_dependency.isWrappedDependencySource(fieldParser)) {
						const pendingState = fieldParser[require_dependency.wrappedDependencySourceMarker];
						const completed = fieldParser.complete([pendingState]);
						if (require_dependency.isDependencySourceState(completed)) {
							preCompletedState[fieldKey] = completed;
							preCompletedKeys.add(fieldKey);
						} else preCompletedState[fieldKey] = fieldState;
					} else preCompletedState[fieldKey] = fieldState;
				}
				const resolvedState = resolveDeferredParseStates(preCompletedState);
				const result = {};
				for (const field of parserKeys) {
					const fieldKey = field;
					const fieldResolvedState = resolvedState[fieldKey];
					const fieldParser = parsers[field];
					if (require_dependency.isDependencySourceState(fieldResolvedState) && preCompletedKeys.has(fieldKey)) {
						const depResult = fieldResolvedState.result;
						if (depResult.success) result[fieldKey] = depResult.value;
						else return {
							success: false,
							error: depResult.error
						};
						continue;
					}
					const valueResult = fieldParser.complete(fieldResolvedState);
					if (valueResult.success) result[fieldKey] = valueResult.value;
					else return {
						success: false,
						error: valueResult.error
					};
				}
				return {
					success: true,
					value: result
				};
			}, async () => {
				const preCompletedState = {};
				const preCompletedKeys = /* @__PURE__ */ new Set();
				for (const field of parserKeys) {
					const fieldKey = field;
					const fieldState = state[fieldKey];
					const fieldParser = parsers[field];
					if (Array.isArray(fieldState) && fieldState.length === 1 && require_dependency.isPendingDependencySourceState(fieldState[0])) {
						const completed = await fieldParser.complete(fieldState);
						preCompletedState[fieldKey] = completed;
						preCompletedKeys.add(fieldKey);
					} else if (fieldState === void 0 && require_dependency.isPendingDependencySourceState(fieldParser.initialState)) {
						const completed = await fieldParser.complete([fieldParser.initialState]);
						preCompletedState[fieldKey] = completed;
						preCompletedKeys.add(fieldKey);
					} else if (fieldState === void 0 && require_dependency.isWrappedDependencySource(fieldParser)) {
						const pendingState = fieldParser[require_dependency.wrappedDependencySourceMarker];
						const completed = await fieldParser.complete([pendingState]);
						if (require_dependency.isDependencySourceState(completed)) {
							preCompletedState[fieldKey] = completed;
							preCompletedKeys.add(fieldKey);
						} else preCompletedState[fieldKey] = fieldState;
					} else preCompletedState[fieldKey] = fieldState;
				}
				const resolvedState = await resolveDeferredParseStatesAsync(preCompletedState);
				const result = {};
				for (const field of parserKeys) {
					const fieldKey = field;
					const fieldResolvedState = resolvedState[fieldKey];
					const fieldParser = parsers[field];
					if (require_dependency.isDependencySourceState(fieldResolvedState) && preCompletedKeys.has(fieldKey)) {
						const depResult = fieldResolvedState.result;
						if (depResult.success) result[fieldKey] = depResult.value;
						else return {
							success: false,
							error: depResult.error
						};
						continue;
					}
					const valueResult = await fieldParser.complete(fieldResolvedState);
					if (valueResult.success) result[fieldKey] = valueResult.value;
					else return {
						success: false,
						error: valueResult.error
					};
				}
				return {
					success: true,
					value: result
				};
			});
		},
		suggest(context, prefix) {
			return require_mode_dispatch.dispatchIterableByMode(combinedMode, () => {
				const syncParserPairs = parserPairs;
				return suggestObjectSync(context, prefix, syncParserPairs);
			}, () => suggestObjectAsync(context, prefix, parserPairs));
		},
		getDocFragments(state, defaultValue) {
			const fragments = parserPairs.flatMap(([field, p]) => {
				const fieldState = state.kind === "unavailable" ? { kind: "unavailable" } : {
					kind: "available",
					state: state.state[field]
				};
				return p.getDocFragments(fieldState, defaultValue?.[field]).fragments;
			});
			const entries = fragments.filter((d) => d.type === "entry");
			const sections = [];
			for (const fragment of fragments) {
				if (fragment.type !== "section") continue;
				if (fragment.title == null) entries.push(...fragment.entries);
				else sections.push(fragment);
			}
			const section = {
				title: label,
				entries
			};
			sections.push(section);
			return { fragments: sections.map((s) => ({
				...s,
				type: "section"
			})) };
		}
	};
}
function suggestTupleSync(context, prefix, parsers) {
	const suggestions = [];
	const stateArray = context.state;
	for (let i = 0; i < parsers.length; i++) {
		const parser = parsers[i];
		const parserState = stateArray && Array.isArray(stateArray) ? stateArray[i] : parser.initialState;
		const parserSuggestions = parser.suggest({
			...context,
			state: parserState
		}, prefix);
		suggestions.push(...parserSuggestions);
	}
	return require_suggestion.deduplicateSuggestions(suggestions);
}
async function* suggestTupleAsync(context, prefix, parsers) {
	const suggestions = [];
	const stateArray = context.state;
	for (let i = 0; i < parsers.length; i++) {
		const parser = parsers[i];
		const parserState = stateArray && Array.isArray(stateArray) ? stateArray[i] : parser.initialState;
		const parserSuggestions = parser.suggest({
			...context,
			state: parserState
		}, prefix);
		if (parser.$mode === "async") for await (const s of parserSuggestions) suggestions.push(s);
		else suggestions.push(...parserSuggestions);
	}
	yield* require_suggestion.deduplicateSuggestions(suggestions);
}
function tuple(labelOrParsers, maybeParsersOrOptions, maybeOptions) {
	const label = typeof labelOrParsers === "string" ? labelOrParsers : void 0;
	let parsers;
	let options = {};
	if (typeof labelOrParsers === "string") {
		parsers = maybeParsersOrOptions;
		options = maybeOptions ?? {};
	} else {
		parsers = labelOrParsers;
		options = maybeParsersOrOptions ?? {};
	}
	const combinedMode = parsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const syncParsers = parsers;
	if (!options.allowDuplicates) checkDuplicateOptionNames(parsers.map((parser, index) => [String(index), parser.usage]));
	const parseSync = (context) => {
		let currentContext = context;
		const allConsumed = [];
		const matchedParsers = /* @__PURE__ */ new Set();
		while (matchedParsers.size < syncParsers.length) {
			let foundMatch = false;
			let error = {
				consumed: 0,
				error: require_message.message`No remaining parsers could match the input.`
			};
			const stateArray = currentContext.state;
			const remainingParsers = syncParsers.map((parser, index) => [parser, index]).filter(([_, index]) => !matchedParsers.has(index)).sort(([parserA], [parserB]) => parserB.priority - parserA.priority);
			for (const [parser, index] of remainingParsers) {
				const result = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length > 0) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: newStateArray
					};
					allConsumed.push(...result.consumed);
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
			if (!foundMatch) for (const [parser, index] of remainingParsers) {
				const result = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length < 1) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						state: newStateArray
					};
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && result.consumed < 1) {
					matchedParsers.add(index);
					foundMatch = true;
					break;
				}
			}
			if (!foundMatch) return {
				...error,
				success: false
			};
		}
		return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
	};
	const parseAsync = async (context) => {
		let currentContext = context;
		const allConsumed = [];
		const matchedParsers = /* @__PURE__ */ new Set();
		while (matchedParsers.size < parsers.length) {
			let foundMatch = false;
			let error = {
				consumed: 0,
				error: require_message.message`No remaining parsers could match the input.`
			};
			const stateArray = currentContext.state;
			const remainingParsers = parsers.map((parser, index) => [parser, index]).filter(([_, index]) => !matchedParsers.has(index)).sort(([parserA], [parserB]) => parserB.priority - parserA.priority);
			for (const [parser, index] of remainingParsers) {
				const resultOrPromise = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				const result = await resultOrPromise;
				if (result.success && result.consumed.length > 0) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: newStateArray
					};
					allConsumed.push(...result.consumed);
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
			if (!foundMatch) for (const [parser, index] of remainingParsers) {
				const resultOrPromise = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				const result = await resultOrPromise;
				if (result.success && result.consumed.length < 1) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						state: newStateArray
					};
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && result.consumed < 1) {
					matchedParsers.add(index);
					foundMatch = true;
					break;
				}
			}
			if (!foundMatch) return {
				...error,
				success: false
			};
		}
		return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		usage: parsers.toSorted((a, b) => b.priority - a.priority).flatMap((p) => p.usage),
		priority: parsers.length > 0 ? Math.max(...parsers.map((p) => p.priority)) : 0,
		initialState: parsers.map((parser) => parser.initialState),
		parse(context) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => parseSync(context), () => parseAsync(context));
		},
		complete(state) {
			return require_mode_dispatch.dispatchByMode(combinedMode, () => {
				const stateArray = state;
				const preCompletedState = [];
				for (let i = 0; i < syncParsers.length; i++) {
					const elementState = stateArray[i];
					const elementParser = syncParsers[i];
					if (Array.isArray(elementState) && elementState.length === 1 && require_dependency.isPendingDependencySourceState(elementState[0])) {
						const completed = elementParser.complete(elementState);
						preCompletedState[i] = completed;
					} else if (elementState === void 0 && require_dependency.isPendingDependencySourceState(elementParser.initialState)) {
						const completed = elementParser.complete([elementParser.initialState]);
						preCompletedState[i] = completed;
					} else if (elementState === void 0 && require_dependency.isWrappedDependencySource(elementParser)) {
						const pendingState = elementParser[require_dependency.wrappedDependencySourceMarker];
						const completed = elementParser.complete([pendingState]);
						preCompletedState[i] = completed;
					} else preCompletedState[i] = elementState;
				}
				const resolvedState = resolveDeferredParseStates(preCompletedState);
				const resolvedArray = resolvedState;
				const result = [];
				for (let i = 0; i < syncParsers.length; i++) {
					const elementResolvedState = resolvedArray[i];
					const elementParser = syncParsers[i];
					const originalElementState = stateArray[i];
					const wasPreCompletedCase1 = Array.isArray(originalElementState) && originalElementState.length === 1 && require_dependency.isPendingDependencySourceState(originalElementState[0]);
					const wasPreCompletedCase2 = originalElementState === void 0 && require_dependency.isPendingDependencySourceState(elementParser.initialState);
					const wasPreCompletedCase3 = originalElementState === void 0 && require_dependency.isWrappedDependencySource(elementParser);
					if (require_dependency.isDependencySourceState(elementResolvedState) && (wasPreCompletedCase1 || wasPreCompletedCase2 || wasPreCompletedCase3)) {
						const depResult = elementResolvedState.result;
						if (depResult.success) result[i] = depResult.value;
						else return {
							success: false,
							error: depResult.error
						};
						continue;
					}
					const valueResult = elementParser.complete(elementResolvedState);
					if (valueResult.success) result[i] = valueResult.value;
					else return {
						success: false,
						error: valueResult.error
					};
				}
				return {
					success: true,
					value: result
				};
			}, async () => {
				const stateArray = state;
				const preCompletedState = [];
				for (let i = 0; i < parsers.length; i++) {
					const elementState = stateArray[i];
					const elementParser = parsers[i];
					if (Array.isArray(elementState) && elementState.length === 1 && require_dependency.isPendingDependencySourceState(elementState[0])) {
						const completed = await elementParser.complete(elementState);
						preCompletedState[i] = completed;
					} else if (elementState === void 0 && require_dependency.isPendingDependencySourceState(elementParser.initialState)) {
						const completed = await elementParser.complete([elementParser.initialState]);
						preCompletedState[i] = completed;
					} else if (elementState === void 0 && require_dependency.isWrappedDependencySource(elementParser)) {
						const pendingState = elementParser[require_dependency.wrappedDependencySourceMarker];
						const completed = await elementParser.complete([pendingState]);
						preCompletedState[i] = completed;
					} else preCompletedState[i] = elementState;
				}
				const resolvedState = await resolveDeferredParseStatesAsync(preCompletedState);
				const resolvedArray = resolvedState;
				const result = [];
				for (let i = 0; i < parsers.length; i++) {
					const elementResolvedState = resolvedArray[i];
					const elementParser = parsers[i];
					const originalElementState = stateArray[i];
					const wasPreCompletedCase1 = Array.isArray(originalElementState) && originalElementState.length === 1 && require_dependency.isPendingDependencySourceState(originalElementState[0]);
					const wasPreCompletedCase2 = originalElementState === void 0 && require_dependency.isPendingDependencySourceState(elementParser.initialState);
					const wasPreCompletedCase3 = originalElementState === void 0 && require_dependency.isWrappedDependencySource(elementParser);
					if (require_dependency.isDependencySourceState(elementResolvedState) && (wasPreCompletedCase1 || wasPreCompletedCase2 || wasPreCompletedCase3)) {
						const depResult = elementResolvedState.result;
						if (depResult.success) result[i] = depResult.value;
						else return {
							success: false,
							error: depResult.error
						};
						continue;
					}
					const valueResult = await elementParser.complete(elementResolvedState);
					if (valueResult.success) result[i] = valueResult.value;
					else return {
						success: false,
						error: valueResult.error
					};
				}
				return {
					success: true,
					value: result
				};
			});
		},
		suggest(context, prefix) {
			return require_mode_dispatch.dispatchIterableByMode(combinedMode, () => suggestTupleSync(context, prefix, syncParsers), () => suggestTupleAsync(context, prefix, parsers));
		},
		getDocFragments(state, defaultValue) {
			const fragments = syncParsers.flatMap((p, i) => {
				const indexState = state.kind === "unavailable" ? { kind: "unavailable" } : {
					kind: "available",
					state: state.state[i]
				};
				return p.getDocFragments(indexState, defaultValue?.[i]).fragments;
			});
			const entries = fragments.filter((d) => d.type === "entry");
			const sections = [];
			for (const fragment of fragments) {
				if (fragment.type !== "section") continue;
				if (fragment.title == null) entries.push(...fragment.entries);
				else sections.push(fragment);
			}
			const section = {
				title: label,
				entries
			};
			sections.push(section);
			return { fragments: sections.map((s) => ({
				...s,
				type: "section"
			})) };
		},
		[Symbol.for("Deno.customInspect")]() {
			const parsersStr = parsers.length === 1 ? `[1 parser]` : `[${parsers.length} parsers]`;
			return label ? `tuple(${JSON.stringify(label)}, ${parsersStr})` : `tuple(${parsersStr})`;
		}
	};
}
function merge(...args) {
	const label = typeof args[0] === "string" ? args[0] : void 0;
	const lastArg = args[args.length - 1];
	const options = lastArg && typeof lastArg === "object" && !("parse" in lastArg) && !("complete" in lastArg) ? lastArg : {};
	const startIndex = typeof args[0] === "string" ? 1 : 0;
	const endIndex = lastArg && typeof lastArg === "object" && !("parse" in lastArg) && !("complete" in lastArg) ? args.length - 1 : args.length;
	const rawParsers = args.slice(startIndex, endIndex);
	const combinedMode = rawParsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const isAsync = combinedMode === "async";
	const syncRawParsers = rawParsers;
	const withIndex = rawParsers.map((p, i) => [p, i]);
	const sorted = withIndex.toSorted(([a], [b]) => b.priority - a.priority);
	const parsers = sorted.map(([p]) => p);
	const syncWithIndex = syncRawParsers.map((p, i) => [p, i]);
	const syncSorted = syncWithIndex.toSorted(([a], [b]) => b.priority - a.priority);
	const syncParsers = syncSorted.map(([p]) => p);
	if (!options.allowDuplicates) checkDuplicateOptionNames(sorted.map(([parser, originalIndex]) => [String(originalIndex), parser.usage]));
	const initialState = {};
	for (let i = 0; i < parsers.length; i++) {
		const parser = parsers[i];
		if (parser.initialState === void 0) initialState[`__parser_${i}`] = void 0;
		else if (parser.initialState && typeof parser.initialState === "object") for (const field in parser.initialState) initialState[field] = parser.initialState[field];
	}
	const extractParserState = (parser, context, index) => {
		if (parser.initialState === void 0) {
			const key = `__parser_${index}`;
			if (context.state && typeof context.state === "object" && key in context.state) return context.state[key];
			return void 0;
		} else if (parser.initialState && typeof parser.initialState === "object") {
			if (context.state && typeof context.state === "object") {
				const extractedState = {};
				for (const field in parser.initialState) extractedState[field] = field in context.state ? context.state[field] : parser.initialState[field];
				return extractedState;
			}
			return parser.initialState;
		}
		return parser.initialState;
	};
	const mergeResultState = (parser, context, result, index) => {
		if (parser.initialState === void 0) {
			const key = `__parser_${index}`;
			if (result.success) {
				if (result.consumed.length > 0 || result.next.state !== void 0) return {
					...context.state,
					[key]: result.next.state
				};
			}
			return { ...context.state };
		}
		return result.success ? {
			...context.state,
			...result.next.state
		} : { ...context.state };
	};
	const parseSync = (context) => {
		let currentContext = context;
		let zeroConsumedSuccess = null;
		for (let i = 0; i < syncParsers.length; i++) {
			const parser = syncParsers[i];
			const parserState = extractParserState(parser, currentContext, i);
			const result = parser.parse({
				...currentContext,
				state: parserState
			});
			if (result.success) {
				const newState = mergeResultState(parser, currentContext, result, i);
				const newContext = {
					...currentContext,
					buffer: result.next.buffer,
					optionsTerminated: result.next.optionsTerminated,
					state: newState
				};
				if (result.consumed.length > 0) return {
					success: true,
					next: newContext,
					consumed: result.consumed
				};
				currentContext = newContext;
				if (zeroConsumedSuccess === null) zeroConsumedSuccess = {
					context: newContext,
					consumed: []
				};
				else zeroConsumedSuccess.context = newContext;
			} else if (result.consumed < 1) continue;
			else return result;
		}
		if (zeroConsumedSuccess !== null) return {
			success: true,
			next: zeroConsumedSuccess.context,
			consumed: zeroConsumedSuccess.consumed
		};
		return {
			success: false,
			consumed: 0,
			error: require_message.message`No matching option or argument found.`
		};
	};
	const parseAsync = async (context) => {
		let currentContext = context;
		let zeroConsumedSuccess = null;
		for (let i = 0; i < parsers.length; i++) {
			const parser = parsers[i];
			const parserState = extractParserState(parser, currentContext, i);
			const resultOrPromise = parser.parse({
				...currentContext,
				state: parserState
			});
			const result = await resultOrPromise;
			if (result.success) {
				const newState = mergeResultState(parser, currentContext, result, i);
				const newContext = {
					...currentContext,
					buffer: result.next.buffer,
					optionsTerminated: result.next.optionsTerminated,
					state: newState
				};
				if (result.consumed.length > 0) return {
					success: true,
					next: newContext,
					consumed: result.consumed
				};
				currentContext = newContext;
				if (zeroConsumedSuccess === null) zeroConsumedSuccess = {
					context: newContext,
					consumed: []
				};
				else zeroConsumedSuccess.context = newContext;
			} else if (result.consumed < 1) continue;
			else return result;
		}
		if (zeroConsumedSuccess !== null) return {
			success: true,
			next: zeroConsumedSuccess.context,
			consumed: zeroConsumedSuccess.consumed
		};
		return {
			success: false,
			consumed: 0,
			error: require_message.message`No matching option or argument found.`
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: Math.max(...parsers.map((p) => p.priority)),
		usage: parsers.flatMap((p) => p.usage),
		initialState,
		parse(context) {
			if (isAsync) return parseAsync(context);
			return parseSync(context);
		},
		complete(state) {
			const extractCompleteState = (parser, resolvedState, index) => {
				if (parser.initialState === void 0) {
					const key = `__parser_${index}`;
					if (resolvedState && typeof resolvedState === "object" && key in resolvedState) return resolvedState[key];
					return void 0;
				} else if (parser.initialState && typeof parser.initialState === "object") {
					if (resolvedState && typeof resolvedState === "object") {
						const extractedState = {};
						for (const field in parser.initialState) extractedState[field] = field in resolvedState ? resolvedState[field] : parser.initialState[field];
						return extractedState;
					}
					return parser.initialState;
				}
				return parser.initialState;
			};
			if (!isAsync) {
				const resolvedState = resolveDeferredParseStates(state);
				const object$1 = {};
				for (let i = 0; i < syncParsers.length; i++) {
					const parser = syncParsers[i];
					const parserState = extractCompleteState(parser, resolvedState, i);
					const result = parser.complete(parserState);
					if (!result.success) return result;
					for (const field in result.value) object$1[field] = result.value[field];
				}
				return {
					success: true,
					value: object$1
				};
			}
			return (async () => {
				const resolvedState = await resolveDeferredParseStatesAsync(state);
				const object$1 = {};
				for (let i = 0; i < parsers.length; i++) {
					const parser = parsers[i];
					const parserState = extractCompleteState(parser, resolvedState, i);
					const result = await parser.complete(parserState);
					if (!result.success) return result;
					for (const field in result.value) object$1[field] = result.value[field];
				}
				return {
					success: true,
					value: object$1
				};
			})();
		},
		suggest(context, prefix) {
			const extractState = (p, i) => {
				if (p.initialState === void 0) {
					const key = `__parser_${i}`;
					if (context.state && typeof context.state === "object" && key in context.state) return context.state[key];
					return void 0;
				} else if (p.initialState && typeof p.initialState === "object") {
					if (context.state && typeof context.state === "object") {
						const extractedState = {};
						for (const field in p.initialState) extractedState[field] = field in context.state ? context.state[field] : p.initialState[field];
						return extractedState;
					}
					return p.initialState;
				}
				return p.initialState;
			};
			if (isAsync) return async function* () {
				const suggestions = [];
				for (let i = 0; i < parsers.length; i++) {
					const parser = parsers[i];
					const parserState = extractState(parser, i);
					const parserSuggestions = parser.suggest({
						...context,
						state: parserState
					}, prefix);
					if (parser.$mode === "async") for await (const s of parserSuggestions) suggestions.push(s);
					else suggestions.push(...parserSuggestions);
				}
				yield* require_suggestion.deduplicateSuggestions(suggestions);
			}();
			return function* () {
				const suggestions = [];
				for (let i = 0; i < syncParsers.length; i++) {
					const parser = syncParsers[i];
					const parserState = extractState(parser, i);
					const parserSuggestions = parser.suggest({
						...context,
						state: parserState
					}, prefix);
					suggestions.push(...parserSuggestions);
				}
				yield* require_suggestion.deduplicateSuggestions(suggestions);
			}();
		},
		getDocFragments(state, _defaultValue) {
			let brief;
			let description;
			let footer;
			const fragments = parsers.flatMap((p, i) => {
				let parserState;
				if (p.initialState === void 0) {
					const key = `__parser_${i}`;
					if (state.kind === "available" && state.state && typeof state.state === "object" && key in state.state) parserState = {
						kind: "available",
						state: state.state[key]
					};
					else parserState = { kind: "unavailable" };
				} else parserState = state.kind === "unavailable" ? { kind: "unavailable" } : {
					kind: "available",
					state: state.state
				};
				const docFragments = p.getDocFragments(parserState, void 0);
				brief ??= docFragments.brief;
				description ??= docFragments.description;
				footer ??= docFragments.footer;
				return docFragments.fragments;
			});
			const entries = fragments.filter((f) => f.type === "entry");
			const sections = [];
			for (const fragment of fragments) {
				if (fragment.type !== "section") continue;
				if (fragment.title == null) entries.push(...fragment.entries);
				else sections.push(fragment);
			}
			if (label) {
				const labeledSection = {
					title: label,
					entries
				};
				sections.push(labeledSection);
				return {
					brief,
					description,
					footer,
					fragments: sections.map((s) => ({
						...s,
						type: "section"
					}))
				};
			}
			return {
				brief,
				description,
				footer,
				fragments: [...sections.map((s) => ({
					...s,
					type: "section"
				})), {
					type: "section",
					entries
				}]
			};
		}
	};
}
function concat(...parsers) {
	const combinedMode = parsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const isAsync = combinedMode === "async";
	const syncParsers = parsers;
	const initialState = parsers.map((parser) => parser.initialState);
	const parseSync = (context) => {
		let currentContext = context;
		const allConsumed = [];
		const matchedParsers = /* @__PURE__ */ new Set();
		while (matchedParsers.size < syncParsers.length) {
			let foundMatch = false;
			let error = {
				consumed: 0,
				error: require_message.message`No remaining parsers could match the input.`
			};
			const stateArray = currentContext.state;
			const remainingParsers = syncParsers.map((parser, index) => [parser, index]).filter(([_, index]) => !matchedParsers.has(index)).sort(([parserA], [parserB]) => parserB.priority - parserA.priority);
			for (const [parser, index] of remainingParsers) {
				const result = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length > 0) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: newStateArray
					};
					allConsumed.push(...result.consumed);
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
			if (!foundMatch) for (const [parser, index] of remainingParsers) {
				const result = parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length < 1) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						state: newStateArray
					};
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && result.consumed < 1) {
					matchedParsers.add(index);
					foundMatch = true;
					break;
				}
			}
			if (!foundMatch) return {
				...error,
				success: false
			};
		}
		return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
	};
	const parseAsync = async (context) => {
		let currentContext = context;
		const allConsumed = [];
		const matchedParsers = /* @__PURE__ */ new Set();
		while (matchedParsers.size < parsers.length) {
			let foundMatch = false;
			let error = {
				consumed: 0,
				error: require_message.message`No remaining parsers could match the input.`
			};
			const stateArray = currentContext.state;
			const remainingParsers = parsers.map((parser, index) => [parser, index]).filter(([_, index]) => !matchedParsers.has(index)).sort(([parserA], [parserB]) => parserB.priority - parserA.priority);
			for (const [parser, index] of remainingParsers) {
				const result = await parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length > 0) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						buffer: result.next.buffer,
						optionsTerminated: result.next.optionsTerminated,
						state: newStateArray
					};
					allConsumed.push(...result.consumed);
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && error.consumed < result.consumed) error = result;
			}
			if (!foundMatch) for (const [parser, index] of remainingParsers) {
				const result = await parser.parse({
					...currentContext,
					state: stateArray[index]
				});
				if (result.success && result.consumed.length < 1) {
					const newStateArray = stateArray.map((s, idx) => idx === index ? result.next.state : s);
					currentContext = {
						...currentContext,
						state: newStateArray
					};
					matchedParsers.add(index);
					foundMatch = true;
					break;
				} else if (!result.success && result.consumed < 1) {
					matchedParsers.add(index);
					foundMatch = true;
					break;
				}
			}
			if (!foundMatch) return {
				...error,
				success: false
			};
		}
		return {
			success: true,
			next: currentContext,
			consumed: allConsumed
		};
	};
	const completeSync = (state) => {
		const stateArray = state;
		const combinedState = {};
		for (let i = 0; i < stateArray.length; i++) combinedState[i] = stateArray[i];
		const resolvedCombinedState = resolveDeferredParseStates(combinedState);
		const results = [];
		for (let i = 0; i < syncParsers.length; i++) {
			const parser = syncParsers[i];
			const parserState = resolvedCombinedState[i];
			const result = parser.complete(parserState);
			if (!result.success) return result;
			if (Array.isArray(result.value)) results.push(...result.value);
			else results.push(result.value);
		}
		return {
			success: true,
			value: results
		};
	};
	const completeAsync = async (state) => {
		const stateArray = state;
		const combinedState = {};
		for (let i = 0; i < stateArray.length; i++) combinedState[i] = stateArray[i];
		const resolvedCombinedState = await resolveDeferredParseStatesAsync(combinedState);
		const results = [];
		for (let i = 0; i < parsers.length; i++) {
			const parser = parsers[i];
			const parserState = resolvedCombinedState[i];
			const result = await parser.complete(parserState);
			if (!result.success) return result;
			if (Array.isArray(result.value)) results.push(...result.value);
			else results.push(result.value);
		}
		return {
			success: true,
			value: results
		};
	};
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: parsers.length > 0 ? Math.max(...parsers.map((p) => p.priority)) : 0,
		usage: parsers.flatMap((p) => p.usage),
		initialState,
		parse(context) {
			if (isAsync) return parseAsync(context);
			return parseSync(context);
		},
		complete(state) {
			if (isAsync) return completeAsync(state);
			return completeSync(state);
		},
		suggest(context, prefix) {
			const stateArray = context.state;
			if (isAsync) return async function* () {
				const suggestions = [];
				for (let i = 0; i < parsers.length; i++) {
					const parser = parsers[i];
					const parserState = stateArray && Array.isArray(stateArray) ? stateArray[i] : parser.initialState;
					const parserSuggestions = parser.suggest({
						...context,
						state: parserState
					}, prefix);
					if (parser.$mode === "async") for await (const s of parserSuggestions) suggestions.push(s);
					else suggestions.push(...parserSuggestions);
				}
				yield* require_suggestion.deduplicateSuggestions(suggestions);
			}();
			return function* () {
				const suggestions = [];
				for (let i = 0; i < syncParsers.length; i++) {
					const parser = syncParsers[i];
					const parserState = stateArray && Array.isArray(stateArray) ? stateArray[i] : parser.initialState;
					const parserSuggestions = parser.suggest({
						...context,
						state: parserState
					}, prefix);
					suggestions.push(...parserSuggestions);
				}
				yield* require_suggestion.deduplicateSuggestions(suggestions);
			}();
		},
		getDocFragments(state, _defaultValue) {
			const fragments = syncParsers.flatMap((p, index) => {
				const indexState = state.kind === "unavailable" ? { kind: "unavailable" } : {
					kind: "available",
					state: state.state[index]
				};
				return p.getDocFragments(indexState, void 0).fragments;
			});
			const entries = fragments.filter((f) => f.type === "entry");
			const sections = [];
			for (const fragment of fragments) {
				if (fragment.type !== "section") continue;
				if (fragment.title == null) entries.push(...fragment.entries);
				else sections.push(fragment);
			}
			const result = [...sections.map((s) => ({
				...s,
				type: "section"
			}))];
			if (entries.length > 0) result.push({
				type: "section",
				entries
			});
			return { fragments: result };
		}
	};
}
/**
* Wraps a parser with a group label for documentation purposes.
*
* The `group()` function is a documentation-only wrapper that applies a label
* to any parser for help text organization. This allows you to use clean code
* structure with combinators like {@link merge} while maintaining well-organized
* help text through group labeling.
*
* The wrapped parser has identical parsing behavior but generates documentation
* fragments wrapped in a labeled section. This is particularly useful when
* combining parsers using {@link merge}—you can wrap the merged result with
* `group()` to add a section header in help output.
*
* @example
* ```typescript
* const apiOptions = merge(
*   object({ endpoint: option("--endpoint", string()) }),
*   object({ timeout: option("--timeout", integer()) })
* );
*
* const groupedApiOptions = group("API Options", apiOptions);
* // Now produces a labeled "API Options" section in help text
* ```
*
* @example
* ```typescript
* // Can be used with any parser, not just merge()
* const verboseGroup = group("Verbosity", object({
*   verbose: option("-v", "--verbose"),
*   quiet: option("-q", "--quiet")
* }));
* ```
*
* @template TValue The value type of the wrapped parser.
* @template TState The state type of the wrapped parser.
* @param label A descriptive label for this parser group, used for
*              documentation and help text organization.
* @param parser The parser to wrap with a group label.
* @returns A new parser that behaves identically to the input parser
*          but generates documentation within a labeled section.
* @since 0.4.0
*/
function group(label, parser) {
	return {
		$mode: parser.$mode,
		$valueType: parser.$valueType,
		$stateType: parser.$stateType,
		priority: parser.priority,
		usage: parser.usage,
		initialState: parser.initialState,
		parse: (context) => parser.parse(context),
		complete: (state) => parser.complete(state),
		suggest: (context, prefix) => parser.suggest(context, prefix),
		getDocFragments: (state, defaultValue) => {
			const { brief, description, footer, fragments } = parser.getDocFragments(state, defaultValue);
			const allEntries = [];
			const titledSections = [];
			for (const fragment of fragments) if (fragment.type === "entry") allEntries.push(fragment);
			else if (fragment.type === "section") if (fragment.title) titledSections.push(fragment);
			else allEntries.push(...fragment.entries);
			const initialFragments = parser.getDocFragments({
				kind: "available",
				state: parser.initialState
			}, void 0);
			const initialCommandNames = /* @__PURE__ */ new Set();
			for (const f of initialFragments.fragments) if (f.type === "entry" && f.term.type === "command") initialCommandNames.add(f.term.name);
			else if (f.type === "section") {
				for (const e of f.entries) if (e.term.type === "command") initialCommandNames.add(e.term.name);
			}
			const initialHasCommands = initialCommandNames.size > 0;
			const currentCommandsAreGroupOwn = allEntries.some((e) => e.term.type === "command" && initialCommandNames.has(e.term.name));
			const applyLabel = !initialHasCommands || currentCommandsAreGroupOwn;
			const labeledSection = applyLabel ? {
				title: label,
				entries: allEntries
			} : { entries: allEntries };
			return {
				brief,
				description,
				footer,
				fragments: [...titledSections.map((s) => ({
					...s,
					type: "section"
				})), {
					type: "section",
					...labeledSection
				}]
			};
		}
	};
}
/**
* Creates a conditional parser that selects different branch parsers based on
* a discriminator option value. This enables discriminated union patterns where
* certain options are only required or available when a specific discriminator
* value is selected.
*
* The result type is a tuple: `[discriminatorValue, branchResult]`
*
* @example
* ```typescript
* // Basic conditional parsing
* const parser = conditional(
*   option("--reporter", choice(["console", "junit"])),
*   {
*     console: object({}),
*     junit: object({ outputFile: option("--output-file", string()) }),
*   },
*   object({}) // default when --reporter is not provided
* );
*
* const result = parse(parser, ["--reporter", "junit", "--output-file", "out.xml"]);
* // result.value = ["junit", { outputFile: "out.xml" }]
*
* // Without --reporter, uses default branch:
* const defaultResult = parse(parser, []);
* // defaultResult.value = [undefined, {}]
* ```
*
* @since 0.8.0
*/
function conditional(discriminator, branches, defaultBranch, options) {
	const branchParsers = Object.entries(branches);
	const allBranchParsers = defaultBranch ? [...branchParsers.map(([_, p]) => p), defaultBranch] : branchParsers.map(([_, p]) => p);
	const combinedMode = discriminator.$mode === "async" || allBranchParsers.some((p) => p.$mode === "async") ? "async" : "sync";
	const isAsync = combinedMode === "async";
	const maxPriority = Math.max(discriminator.priority, ...allBranchParsers.map((p) => p.priority));
	function appendLiteralToUsage(usage$1, literalValue) {
		const result = [];
		for (const term of usage$1) if (term.type === "option" && term.metavar !== void 0) {
			const { metavar: _,...optionWithoutMetavar } = term;
			result.push(optionWithoutMetavar);
			result.push({
				type: "literal",
				value: literalValue
			});
		} else if (term.type === "optional") result.push({
			...term,
			terms: appendLiteralToUsage(term.terms, literalValue)
		});
		else if (term.type === "multiple") result.push({
			...term,
			terms: appendLiteralToUsage(term.terms, literalValue)
		});
		else if (term.type === "exclusive") result.push({
			...term,
			terms: term.terms.map((t) => appendLiteralToUsage(t, literalValue))
		});
		else result.push(term);
		return result;
	}
	const branchUsages = branchParsers.map(([key, p]) => [...appendLiteralToUsage(discriminator.usage, key), ...p.usage]);
	if (defaultBranch) branchUsages.push(defaultBranch.usage);
	const usage = branchUsages.length > 1 ? [{
		type: "exclusive",
		terms: branchUsages
	}] : branchUsages[0] ?? [];
	const initialState = {
		discriminatorState: discriminator.initialState,
		discriminatorValue: void 0,
		selectedBranch: void 0,
		branchState: void 0
	};
	const getNoMatchError$1 = () => {
		const noMatchContext = analyzeNoMatchContext([discriminator, ...allBranchParsers]);
		return options?.errors?.noMatch ? typeof options.errors.noMatch === "function" ? options.errors.noMatch(noMatchContext) : options.errors.noMatch : generateNoMatchError(noMatchContext);
	};
	const parseSync = (context) => {
		const state = context.state ?? initialState;
		const syncDiscriminator = discriminator;
		const syncBranches = branches;
		const syncDefaultBranch = defaultBranch;
		if (state.selectedBranch !== void 0) {
			const branchParser = state.selectedBranch.kind === "default" ? syncDefaultBranch : syncBranches[state.selectedBranch.key];
			const branchResult = branchParser.parse({
				...context,
				state: state.branchState,
				usage: branchParser.usage
			});
			if (branchResult.success) return {
				success: true,
				next: {
					...branchResult.next,
					state: {
						...state,
						branchState: branchResult.next.state
					}
				},
				consumed: branchResult.consumed
			};
			return branchResult;
		}
		const discriminatorResult = syncDiscriminator.parse({
			...context,
			state: state.discriminatorState
		});
		if (discriminatorResult.success && discriminatorResult.consumed.length > 0) {
			const completionResult = syncDiscriminator.complete(discriminatorResult.next.state);
			if (completionResult.success) {
				const value = completionResult.value;
				const branchParser = syncBranches[value];
				if (branchParser) {
					const branchParseResult = branchParser.parse({
						...context,
						buffer: discriminatorResult.next.buffer,
						optionsTerminated: discriminatorResult.next.optionsTerminated,
						state: branchParser.initialState,
						usage: branchParser.usage
					});
					if (branchParseResult.success) return {
						success: true,
						next: {
							...branchParseResult.next,
							state: {
								discriminatorState: discriminatorResult.next.state,
								discriminatorValue: value,
								selectedBranch: {
									kind: "branch",
									key: value
								},
								branchState: branchParseResult.next.state
							}
						},
						consumed: [...discriminatorResult.consumed, ...branchParseResult.consumed]
					};
					return {
						success: true,
						next: {
							...discriminatorResult.next,
							state: {
								discriminatorState: discriminatorResult.next.state,
								discriminatorValue: value,
								selectedBranch: {
									kind: "branch",
									key: value
								},
								branchState: branchParser.initialState
							}
						},
						consumed: discriminatorResult.consumed
					};
				}
			}
		}
		if (syncDefaultBranch !== void 0) {
			const defaultResult = syncDefaultBranch.parse({
				...context,
				state: state.branchState ?? syncDefaultBranch.initialState,
				usage: syncDefaultBranch.usage
			});
			if (defaultResult.success && defaultResult.consumed.length > 0) return {
				success: true,
				next: {
					...defaultResult.next,
					state: {
						...state,
						selectedBranch: { kind: "default" },
						branchState: defaultResult.next.state
					}
				},
				consumed: defaultResult.consumed
			};
		}
		return {
			success: false,
			consumed: 0,
			error: getNoMatchError$1()
		};
	};
	const parseAsync = async (context) => {
		const state = context.state ?? initialState;
		if (state.selectedBranch !== void 0) {
			const branchParser = state.selectedBranch.kind === "default" ? defaultBranch : branches[state.selectedBranch.key];
			const branchResult = await branchParser.parse({
				...context,
				state: state.branchState,
				usage: branchParser.usage
			});
			if (branchResult.success) return {
				success: true,
				next: {
					...branchResult.next,
					state: {
						...state,
						branchState: branchResult.next.state
					}
				},
				consumed: branchResult.consumed
			};
			return branchResult;
		}
		const discriminatorResult = await discriminator.parse({
			...context,
			state: state.discriminatorState
		});
		if (discriminatorResult.success && discriminatorResult.consumed.length > 0) {
			const completionResult = await discriminator.complete(discriminatorResult.next.state);
			if (completionResult.success) {
				const value = completionResult.value;
				const branchParser = branches[value];
				if (branchParser) {
					const branchParseResult = await branchParser.parse({
						...context,
						buffer: discriminatorResult.next.buffer,
						optionsTerminated: discriminatorResult.next.optionsTerminated,
						state: branchParser.initialState,
						usage: branchParser.usage
					});
					if (branchParseResult.success) return {
						success: true,
						next: {
							...branchParseResult.next,
							state: {
								discriminatorState: discriminatorResult.next.state,
								discriminatorValue: value,
								selectedBranch: {
									kind: "branch",
									key: value
								},
								branchState: branchParseResult.next.state
							}
						},
						consumed: [...discriminatorResult.consumed, ...branchParseResult.consumed]
					};
					return {
						success: true,
						next: {
							...discriminatorResult.next,
							state: {
								discriminatorState: discriminatorResult.next.state,
								discriminatorValue: value,
								selectedBranch: {
									kind: "branch",
									key: value
								},
								branchState: branchParser.initialState
							}
						},
						consumed: discriminatorResult.consumed
					};
				}
			}
		}
		if (defaultBranch !== void 0) {
			const defaultResult = await defaultBranch.parse({
				...context,
				state: state.branchState ?? defaultBranch.initialState,
				usage: defaultBranch.usage
			});
			if (defaultResult.success && defaultResult.consumed.length > 0) return {
				success: true,
				next: {
					...defaultResult.next,
					state: {
						...state,
						selectedBranch: { kind: "default" },
						branchState: defaultResult.next.state
					}
				},
				consumed: defaultResult.consumed
			};
		}
		return {
			success: false,
			consumed: 0,
			error: getNoMatchError$1()
		};
	};
	const completeSync = (state) => {
		const syncDiscriminator = discriminator;
		const syncDefaultBranch = defaultBranch;
		const syncBranches = branches;
		if (state.selectedBranch === void 0) {
			if (syncDefaultBranch !== void 0) {
				const branchState = state.branchState ?? syncDefaultBranch.initialState;
				const defaultResult = syncDefaultBranch.complete(branchState);
				if (!defaultResult.success) return defaultResult;
				return {
					success: true,
					value: [void 0, defaultResult.value]
				};
			}
			return {
				success: false,
				error: require_message.message`Missing required discriminator option.`
			};
		}
		const branchParser = state.selectedBranch.kind === "default" ? syncDefaultBranch : syncBranches[state.selectedBranch.key];
		const discriminatorCompleteResult = syncDiscriminator.complete(state.discriminatorState);
		const combinedState = {
			_discriminator: state.discriminatorState,
			_branch: state.branchState
		};
		const resolvedCombinedState = resolveDeferredParseStates(combinedState);
		const resolvedBranchState = resolvedCombinedState._branch;
		const branchResult = branchParser.complete(resolvedBranchState);
		if (!branchResult.success) {
			if (state.discriminatorValue !== void 0 && options?.errors?.branchError) return {
				success: false,
				error: options.errors.branchError(state.discriminatorValue, branchResult.error)
			};
			return branchResult;
		}
		let discriminatorValue;
		if (state.selectedBranch.kind === "default") discriminatorValue = void 0;
		else if (require_dependency.isDependencySourceState(discriminatorCompleteResult)) discriminatorValue = discriminatorCompleteResult.result.success ? discriminatorCompleteResult.result.value : state.selectedBranch.key;
		else if (discriminatorCompleteResult.success) discriminatorValue = discriminatorCompleteResult.value;
		else discriminatorValue = state.selectedBranch.key;
		return {
			success: true,
			value: [discriminatorValue, branchResult.value]
		};
	};
	const completeAsync = async (state) => {
		if (state.selectedBranch === void 0) {
			if (defaultBranch !== void 0) {
				const branchState = state.branchState ?? defaultBranch.initialState;
				const defaultResult = await defaultBranch.complete(branchState);
				if (!defaultResult.success) return defaultResult;
				return {
					success: true,
					value: [void 0, defaultResult.value]
				};
			}
			return {
				success: false,
				error: require_message.message`Missing required discriminator option.`
			};
		}
		const branchParser = state.selectedBranch.kind === "default" ? defaultBranch : branches[state.selectedBranch.key];
		const discriminatorCompleteResult = await discriminator.complete(state.discriminatorState);
		const combinedState = {
			_discriminator: state.discriminatorState,
			_branch: state.branchState
		};
		const resolvedCombinedState = await resolveDeferredParseStatesAsync(combinedState);
		const resolvedBranchState = resolvedCombinedState._branch;
		const branchResult = await branchParser.complete(resolvedBranchState);
		if (!branchResult.success) {
			if (state.discriminatorValue !== void 0 && options?.errors?.branchError) return {
				success: false,
				error: options.errors.branchError(state.discriminatorValue, branchResult.error)
			};
			return branchResult;
		}
		let discriminatorValue;
		if (state.selectedBranch.kind === "default") discriminatorValue = void 0;
		else if (require_dependency.isDependencySourceState(discriminatorCompleteResult)) discriminatorValue = discriminatorCompleteResult.result.success ? discriminatorCompleteResult.result.value : state.selectedBranch.key;
		else if (discriminatorCompleteResult.success) discriminatorValue = discriminatorCompleteResult.value;
		else discriminatorValue = state.selectedBranch.key;
		return {
			success: true,
			value: [discriminatorValue, branchResult.value]
		};
	};
	function* suggestSync(context, prefix) {
		const state = context.state ?? initialState;
		const syncDiscriminator = discriminator;
		const syncBranches = branches;
		const syncDefaultBranch = defaultBranch;
		if (state.selectedBranch === void 0) {
			yield* syncDiscriminator.suggest({
				...context,
				state: state.discriminatorState
			}, prefix);
			if (syncDefaultBranch !== void 0) yield* syncDefaultBranch.suggest({
				...context,
				state: state.branchState ?? syncDefaultBranch.initialState
			}, prefix);
		} else {
			const branchParser = state.selectedBranch.kind === "default" ? syncDefaultBranch : syncBranches[state.selectedBranch.key];
			yield* branchParser.suggest({
				...context,
				state: state.branchState
			}, prefix);
		}
	}
	async function* suggestAsync(context, prefix) {
		const state = context.state ?? initialState;
		if (state.selectedBranch === void 0) {
			yield* discriminator.suggest({
				...context,
				state: state.discriminatorState
			}, prefix);
			if (defaultBranch !== void 0) yield* defaultBranch.suggest({
				...context,
				state: state.branchState ?? defaultBranch.initialState
			}, prefix);
		} else {
			const branchParser = state.selectedBranch.kind === "default" ? defaultBranch : branches[state.selectedBranch.key];
			yield* branchParser.suggest({
				...context,
				state: state.branchState
			}, prefix);
		}
	}
	return {
		$mode: combinedMode,
		$valueType: [],
		$stateType: [],
		priority: maxPriority,
		usage,
		initialState,
		parse(context) {
			if (isAsync) return parseAsync(context);
			return parseSync(context);
		},
		complete(state) {
			if (isAsync) return completeAsync(state);
			return completeSync(state);
		},
		suggest(context, prefix) {
			if (isAsync) return suggestAsync(context, prefix);
			return suggestSync(context, prefix);
		},
		getDocFragments(_state, _defaultValue) {
			const fragments = [];
			const discriminatorFragments = discriminator.getDocFragments({ kind: "unavailable" }, void 0);
			fragments.push(...discriminatorFragments.fragments);
			for (const [key, branchParser] of branchParsers) {
				const branchFragments = branchParser.getDocFragments({ kind: "unavailable" }, void 0);
				const entries = branchFragments.fragments.filter((f) => f.type === "entry");
				for (const fragment of branchFragments.fragments) if (fragment.type === "section") entries.push(...fragment.entries);
				if (entries.length > 0) fragments.push({
					type: "section",
					title: `Options when ${key}`,
					entries
				});
			}
			if (defaultBranch !== void 0) {
				const defaultFragments = defaultBranch.getDocFragments({ kind: "unavailable" }, void 0);
				const entries = defaultFragments.fragments.filter((f) => f.type === "entry");
				for (const fragment of defaultFragments.fragments) if (fragment.type === "section") entries.push(...fragment.entries);
				if (entries.length > 0) fragments.push({
					type: "section",
					title: "Default options",
					entries
				});
			}
			return { fragments };
		}
	};
}

//#endregion
exports.DuplicateOptionError = DuplicateOptionError;
exports.concat = concat;
exports.conditional = conditional;
exports.group = group;
exports.longestMatch = longestMatch;
exports.merge = merge;
exports.object = object;
exports.or = or;
exports.tuple = tuple;
