// @optique/core/dist/doc.js
import { formatMessage, text } from "./message.js";
import { formatUsage, formatUsageTerm } from "./usage.js";

//#region src/doc.ts
/**
* Formats a documentation page into a human-readable string.
*
* This function takes a structured {@link DocPage} and converts it into
* a formatted string suitable for display in terminals or documentation.
* The formatting includes proper indentation, alignment, and optional
* color support.
*
* @param programName The name of the program, used in usage lines
* @param page The documentation page to format
* @param options Formatting options to customize the output
* @returns A formatted string representation of the documentation page
*
* @example
* ```typescript
* const page: DocPage = {
*   brief: "A CLI tool",
*   usage: [{ type: "literal", value: "myapp" }],
*   sections: [{
*     title: "Options",
*     entries: [{
*       term: { type: "option", short: "-v", long: "--verbose" },
*       description: "Enable verbose output"
*     }]
*   }]
* };
*
* const formatted = formatDocPage("myapp", page, { colors: true });
* console.log(formatted);
* ```
*/
function formatDocPage(programName, page, options = {}) {
	const termIndent = options.termIndent ?? 2;
	const termWidth = options.termWidth ?? 26;
	let output = "";
	if (page.brief != null) {
		output += formatMessage(page.brief, {
			colors: options.colors,
			maxWidth: options.maxWidth,
			quotes: !options.colors
		});
		output += "\n";
	}
	if (page.usage != null) {
		const usageLabel = options.colors ? "\x1B[1;2mUsage:\x1B[0m " : "Usage: ";
		output += usageLabel;
		output += indentLines(formatUsage(programName, page.usage, {
			colors: options.colors,
			maxWidth: options.maxWidth == null ? void 0 : options.maxWidth - 7,
			expandCommands: true
		}), 7);
		output += "\n";
	}
	if (page.description != null) {
		output += "\n";
		output += formatMessage(page.description, {
			colors: options.colors,
			maxWidth: options.maxWidth,
			quotes: !options.colors
		});
		output += "\n";
	}
	const sections = page.sections.toSorted((a, b) => a.title == null && b.title == null ? 0 : a.title == null ? -1 : 1);
	for (const section of sections) {
		if (section.entries.length < 1) continue;
		output += "\n";
		if (section.title != null) {
			const sectionLabel = options.colors ? `\x1b[1;2m${section.title}:\x1b[0m\n` : `${section.title}:\n`;
			output += sectionLabel;
		}
		for (const entry of section.entries) {
			const term = formatUsageTerm(entry.term, {
				colors: options.colors,
				optionsSeparator: ", ",
				maxWidth: options.maxWidth == null ? void 0 : options.maxWidth - termIndent
			});
			const descColumnWidth = options.maxWidth == null ? void 0 : options.maxWidth - termIndent - termWidth - 2;
			const termVisibleWidth = lastLineVisibleLength(term);
			const extraTermOffset = descColumnWidth != null ? Math.max(0, termVisibleWidth - termWidth) : 0;
			const currentExtraOffset = () => description.includes("\n") ? 0 : extraTermOffset;
			const descFormatOptions = {
				colors: options.colors,
				quotes: !options.colors,
				maxWidth: descColumnWidth,
				startWidth: extraTermOffset > 0 ? extraTermOffset : void 0
			};
			let description = entry.description == null ? "" : formatMessage(entry.description, descFormatOptions);
			if (options.showDefault && entry.default != null) {
				const prefix = typeof options.showDefault === "object" ? options.showDefault.prefix ?? " [" : " [";
				const suffix = typeof options.showDefault === "object" ? options.showDefault.suffix ?? "]" : "]";
				let defaultStartWidth;
				if (descColumnWidth != null) {
					const lastW = lastLineVisibleLength(description);
					const effectiveLastW = lastW + currentExtraOffset();
					if (effectiveLastW + prefix.length >= descColumnWidth) {
						description += "\n";
						defaultStartWidth = prefix.length;
					} else defaultStartWidth = effectiveLastW + prefix.length;
				}
				const defaultFormatOptions = {
					colors: options.colors ? { resetSuffix: "\x1B[2m" } : false,
					quotes: !options.colors,
					maxWidth: descColumnWidth == null ? void 0 : descColumnWidth - suffix.length,
					startWidth: defaultStartWidth
				};
				const defaultContent = formatMessage(entry.default, defaultFormatOptions);
				const defaultText = `${prefix}${defaultContent}${suffix}`;
				const formattedDefault = options.colors ? `\x1b[2m${defaultText}\x1b[0m` : defaultText;
				description += formattedDefault;
			}
			if (options.showChoices && entry.choices != null) {
				const prefix = typeof options.showChoices === "object" ? options.showChoices.prefix ?? " (" : " (";
				const suffix = typeof options.showChoices === "object" ? options.showChoices.suffix ?? ")" : ")";
				const label = typeof options.showChoices === "object" ? options.showChoices.label ?? "choices: " : "choices: ";
				const maxItems = typeof options.showChoices === "object" ? options.showChoices.maxItems ?? 8 : 8;
				const terms = Array.isArray(entry.choices) ? entry.choices : [];
				let truncatedTerms = terms;
				let truncated = false;
				if (maxItems < Infinity) {
					let valueCount = 0;
					let cutIndex = terms.length;
					for (let i = 0; i < terms.length; i++) if (terms[i].type === "value") {
						valueCount++;
						if (valueCount > maxItems) {
							cutIndex = i > 0 && terms[i - 1].type === "text" ? i - 1 : i;
							truncated = true;
							break;
						}
					}
					if (truncated) truncatedTerms = [...terms.slice(0, cutIndex), text(", ...")];
				}
				let choicesStartWidth;
				if (descColumnWidth != null) {
					const lastW = lastLineVisibleLength(description);
					const effectiveLastW = lastW + currentExtraOffset();
					const prefixLabelLen = prefix.length + label.length;
					if (effectiveLastW + prefixLabelLen >= descColumnWidth) {
						description += "\n";
						choicesStartWidth = prefixLabelLen;
					} else choicesStartWidth = effectiveLastW + prefixLabelLen;
				}
				const choicesFormatOptions = {
					colors: options.colors ? { resetSuffix: "\x1B[2m" } : false,
					quotes: false,
					maxWidth: descColumnWidth == null ? void 0 : descColumnWidth - suffix.length,
					startWidth: choicesStartWidth
				};
				const choicesDisplay = formatMessage(truncatedTerms, choicesFormatOptions);
				const choicesText = `${prefix}${label}${choicesDisplay}${suffix}`;
				const formattedChoices = options.colors ? `\x1b[2m${choicesText}\x1b[0m` : choicesText;
				description += formattedChoices;
			}
			output += `${" ".repeat(termIndent)}${ansiAwareRightPad(term, termWidth)}  ${description === "" ? "" : indentLines(description, termIndent + termWidth + 2)}\n`;
		}
	}
	if (page.examples != null) {
		output += "\n";
		const examplesLabel = options.colors ? "\x1B[1;2mExamples:\x1B[0m\n" : "Examples:\n";
		output += examplesLabel;
		const examplesContent = formatMessage(page.examples, {
			colors: options.colors,
			maxWidth: options.maxWidth == null ? void 0 : options.maxWidth - 2,
			quotes: !options.colors
		});
		output += "  " + indentLines(examplesContent, 2);
		output += "\n";
	}
	if (page.author != null) {
		output += "\n";
		const authorLabel = options.colors ? "\x1B[1;2mAuthor:\x1B[0m\n" : "Author:\n";
		output += authorLabel;
		const authorContent = formatMessage(page.author, {
			colors: options.colors,
			maxWidth: options.maxWidth == null ? void 0 : options.maxWidth - 2,
			quotes: !options.colors
		});
		output += "  " + indentLines(authorContent, 2);
		output += "\n";
	}
	if (page.bugs != null) {
		output += "\n";
		const bugsLabel = options.colors ? "\x1B[1;2mBugs:\x1B[0m\n" : "Bugs:\n";
		output += bugsLabel;
		const bugsContent = formatMessage(page.bugs, {
			colors: options.colors,
			maxWidth: options.maxWidth == null ? void 0 : options.maxWidth - 2,
			quotes: !options.colors
		});
		output += "  " + indentLines(bugsContent, 2);
		output += "\n";
	}
	if (page.footer != null) {
		output += "\n";
		output += formatMessage(page.footer, {
			colors: options.colors,
			maxWidth: options.maxWidth,
			quotes: !options.colors
		});
	}
	return output;
}
function indentLines(text$1, indent) {
	return text$1.split("\n").join("\n" + " ".repeat(indent));
}
const ansiEscapeCodeRegex = /\x1B\[[0-9;]*[a-zA-Z]/g;
function ansiAwareRightPad(text$1, length, char = " ") {
	const strippedText = text$1.replace(ansiEscapeCodeRegex, "");
	if (strippedText.length >= length) return text$1;
	return text$1 + char.repeat(length - strippedText.length);
}
function lastLineVisibleLength(text$1) {
	const lastNewline = text$1.lastIndexOf("\n");
	const lastLine = lastNewline === -1 ? text$1 : text$1.slice(lastNewline + 1);
	return lastLine.replace(ansiEscapeCodeRegex, "").length;
}

//#endregion
export { formatDocPage };
