// @optique/core/dist/facade.cjs
const require_annotations = require('./annotations.cjs');
const require_message = require('./message.cjs');
const require_completion = require('./completion.cjs');
const require_usage = require('./usage.cjs');
const require_constructs = require('./constructs.cjs');
const require_doc = require('./doc.cjs');
const require_modifiers = require('./modifiers.cjs');
const require_valueparser = require('./valueparser.cjs');
const require_primitives = require('./primitives.cjs');
const require_parser = require('./parser.cjs');

//#region src/facade.ts
/**
* Creates help parsers based on the specified mode.
*/
function createHelpParser(mode) {
	const helpCommand = require_primitives.command("help", require_modifiers.multiple(require_primitives.argument(require_valueparser.string({ metavar: "COMMAND" }), { description: require_message.message`Command name to show help for.` })), { description: require_message.message`Show help information.` });
	const helpOption = require_primitives.flag("--help", { description: require_message.message`Show help information.` });
	switch (mode) {
		case "command": return {
			helpCommand,
			helpOption: null
		};
		case "option": return {
			helpCommand: null,
			helpOption
		};
		case "both": return {
			helpCommand,
			helpOption
		};
	}
}
/**
* Creates version parsers based on the specified mode.
*/
function createVersionParser(mode) {
	const versionCommand = require_primitives.command("version", require_constructs.object({}), { description: require_message.message`Show version information.` });
	const versionOption = require_primitives.flag("--version", { description: require_message.message`Show version information.` });
	switch (mode) {
		case "command": return {
			versionCommand,
			versionOption: null
		};
		case "option": return {
			versionCommand: null,
			versionOption
		};
		case "both": return {
			versionCommand,
			versionOption
		};
	}
}
/**
* Creates completion parsers based on the specified mode.
*/
function createCompletionParser(mode, programName, availableShells, name = "both", helpVisibility = name) {
	const shellList = [];
	for (const shell in availableShells) {
		if (shellList.length > 0) shellList.push(require_message.text(", "));
		shellList.push(require_message.value(shell));
	}
	const completionInner = require_constructs.object({
		shell: require_modifiers.optional(require_primitives.argument(require_valueparser.string({ metavar: "SHELL" }), { description: require_message.message`Shell type (${shellList}). Generate completion script when used alone, or provide completions when followed by arguments.` })),
		args: require_modifiers.multiple(require_primitives.argument(require_valueparser.string({ metavar: "ARG" }), { description: require_message.message`Command line arguments for completion suggestions (used by shell integration; you usually don't need to provide this).` }))
	});
	const commandName = name === "plural" ? "completions" : "completion";
	const completionCommandConfig = {
		brief: require_message.message`Generate shell completion script or provide completions.`,
		description: require_message.message`Generate shell completion script or provide completions.`,
		footer: require_message.message`Examples:${require_message.lineBreak()}  Bash:       ${require_message.commandLine(`eval "$(${programName} ${commandName} bash)"`)}${require_message.lineBreak()}  zsh:        ${require_message.commandLine(`eval "$(${programName} ${commandName} zsh)"`)}${require_message.lineBreak()}  fish:       ${require_message.commandLine(`eval "$(${programName} ${commandName} fish)"`)}${require_message.lineBreak()}  PowerShell: ${require_message.commandLine(`${programName} ${commandName} pwsh > ${programName}-completion.ps1; . ./${programName}-completion.ps1`)}${require_message.lineBreak()}  Nushell:    ${require_message.commandLine(`${programName} ${commandName} nu | save ${programName}-completion.nu; source ./${programName}-completion.nu`)}`
	};
	const completionCommands = [];
	const showSingular = helpVisibility === "singular" || helpVisibility === "both";
	const showPlural = helpVisibility === "plural" || helpVisibility === "both";
	if (name === "singular" || name === "both") completionCommands.push(require_primitives.command("completion", completionInner, {
		...completionCommandConfig,
		hidden: !showSingular
	}));
	if (name === "plural" || name === "both") completionCommands.push(require_primitives.command("completions", completionInner, {
		...completionCommandConfig,
		hidden: !showPlural
	}));
	const completionCommand = require_constructs.longestMatch(...completionCommands);
	const completionOptions = [];
	if (name === "singular" || name === "both") completionOptions.push(require_primitives.option("--completion", require_valueparser.string({ metavar: "SHELL" }), {
		description: require_message.message`Generate shell completion script.`,
		hidden: !showSingular
	}));
	if (name === "plural" || name === "both") completionOptions.push(require_primitives.option("--completions", require_valueparser.string({ metavar: "SHELL" }), {
		description: require_message.message`Generate shell completion script.`,
		hidden: !showPlural
	}));
	const completionOption = completionOptions.length === 1 ? completionOptions[0] : require_constructs.longestMatch(completionOptions[0], completionOptions[1]);
	const argsParser = require_modifiers.withDefault(require_modifiers.multiple(require_primitives.argument(require_valueparser.string({ metavar: "ARG" }), { description: require_message.message`Command line arguments for completion suggestions (used by shell integration; you usually don't need to provide this).` })), []);
	const optionParser = require_constructs.object({
		shell: completionOption,
		args: argsParser
	});
	switch (mode) {
		case "command": return {
			completionCommand,
			completionOption: null
		};
		case "option": return {
			completionCommand: null,
			completionOption: optionParser
		};
		case "both": return {
			completionCommand,
			completionOption: optionParser
		};
	}
}
function combineWithHelpVersion(originalParser, helpParsers, versionParsers, completionParsers, groups) {
	const parsers = [];
	if (helpParsers.helpOption) {
		const lenientHelpParser = {
			$mode: "sync",
			$valueType: [],
			$stateType: [],
			priority: 200,
			usage: helpParsers.helpOption.usage,
			initialState: null,
			parse(context) {
				const { buffer, optionsTerminated } = context;
				if (optionsTerminated) return {
					success: false,
					error: require_message.message`Options terminated.`,
					consumed: 0
				};
				let helpFound = false;
				let helpIndex = -1;
				let versionIndex = -1;
				for (let i = 0; i < buffer.length; i++) {
					if (buffer[i] === "--") break;
					if (buffer[i] === "--help") {
						helpFound = true;
						helpIndex = i;
					}
					if (buffer[i] === "--version") versionIndex = i;
				}
				if (helpFound && versionIndex > helpIndex) return {
					success: false,
					error: require_message.message`Version option wins.`,
					consumed: 0
				};
				if (helpFound) {
					const commands = [];
					for (let i = 0; i < helpIndex; i++) {
						const arg = buffer[i];
						if (!arg.startsWith("-")) commands.push(arg);
					}
					return {
						success: true,
						next: {
							...context,
							buffer: [],
							state: {
								help: true,
								version: false,
								commands,
								helpFlag: true
							}
						},
						consumed: buffer.slice(0)
					};
				}
				return {
					success: false,
					error: require_message.message`Flag ${require_message.optionName("--help")} not found.`,
					consumed: 0
				};
			},
			complete(state) {
				return {
					success: true,
					value: state
				};
			},
			*suggest(_context, prefix) {
				if ("--help".startsWith(prefix)) yield {
					kind: "literal",
					text: "--help"
				};
			},
			getDocFragments(state) {
				return helpParsers.helpOption?.getDocFragments(state) ?? { fragments: [] };
			}
		};
		parsers.push(lenientHelpParser);
	}
	if (versionParsers.versionOption) {
		const lenientVersionParser = {
			$mode: "sync",
			$valueType: [],
			$stateType: [],
			priority: 200,
			usage: versionParsers.versionOption.usage,
			initialState: null,
			parse(context) {
				const { buffer, optionsTerminated } = context;
				if (optionsTerminated) return {
					success: false,
					error: require_message.message`Options terminated.`,
					consumed: 0
				};
				let versionFound = false;
				let versionIndex = -1;
				let helpIndex = -1;
				for (let i = 0; i < buffer.length; i++) {
					if (buffer[i] === "--") break;
					if (buffer[i] === "--version") {
						versionFound = true;
						versionIndex = i;
					}
					if (buffer[i] === "--help") helpIndex = i;
				}
				if (versionFound && helpIndex > versionIndex) return {
					success: false,
					error: require_message.message`Help option wins.`,
					consumed: 0
				};
				if (versionFound) return {
					success: true,
					next: {
						...context,
						buffer: [],
						state: {
							help: false,
							version: true,
							versionFlag: true
						}
					},
					consumed: buffer.slice(0)
				};
				return {
					success: false,
					error: require_message.message`Flag ${require_message.optionName("--version")} not found.`,
					consumed: 0
				};
			},
			complete(state) {
				return {
					success: true,
					value: state
				};
			},
			*suggest(_context, prefix) {
				if ("--version".startsWith(prefix)) yield {
					kind: "literal",
					text: "--version"
				};
			},
			getDocFragments(state) {
				return versionParsers.versionOption?.getDocFragments(state) ?? { fragments: [] };
			}
		};
		parsers.push(lenientVersionParser);
	}
	if (versionParsers.versionCommand) {
		const versionParser = require_constructs.object({
			help: require_primitives.constant(false),
			version: require_primitives.constant(true),
			completion: require_primitives.constant(false),
			result: versionParsers.versionCommand,
			helpFlag: helpParsers.helpOption ? require_modifiers.optional(helpParsers.helpOption) : require_primitives.constant(false)
		});
		parsers.push(groups?.versionGroup ? require_constructs.group(groups.versionGroup, versionParser) : versionParser);
	}
	if (completionParsers.completionCommand) {
		const completionParser = require_constructs.object({
			help: require_primitives.constant(false),
			version: require_primitives.constant(false),
			completion: require_primitives.constant(true),
			completionData: completionParsers.completionCommand,
			helpFlag: helpParsers.helpOption ? require_modifiers.optional(helpParsers.helpOption) : require_primitives.constant(false)
		});
		parsers.push(groups?.completionGroup ? require_constructs.group(groups.completionGroup, completionParser) : completionParser);
	}
	if (helpParsers.helpCommand) {
		const helpParser = require_constructs.object({
			help: require_primitives.constant(true),
			version: require_primitives.constant(false),
			completion: require_primitives.constant(false),
			commands: helpParsers.helpCommand
		});
		parsers.push(groups?.helpGroup ? require_constructs.group(groups.helpGroup, helpParser) : helpParser);
	}
	parsers.push(require_constructs.object({
		help: require_primitives.constant(false),
		version: require_primitives.constant(false),
		completion: require_primitives.constant(false),
		result: originalParser
	}));
	const mainParserIndex = parsers.length - 1;
	if (parsers.length === 1) return parsers[0];
	let combined;
	if (parsers.length === 2) combined = require_constructs.longestMatch(parsers[0], parsers[1]);
	else combined = require_constructs.longestMatch(...parsers);
	const topUsage = combined.usage[0];
	if (topUsage?.type === "exclusive" && mainParserIndex > 0) {
		const terms = [...topUsage.terms];
		const [mainTerm] = terms.splice(mainParserIndex, 1);
		const lenientCount = (helpParsers.helpOption ? 1 : 0) + (versionParsers.versionOption ? 1 : 0);
		terms.splice(lenientCount, 0, mainTerm);
		combined = {
			...combined,
			usage: [{
				...topUsage,
				terms
			}]
		};
	}
	return combined;
}
/**
* Classifies the parsing result into a discriminated union for cleaner handling.
*/
function classifyResult(result, args) {
	if (!result.success) return {
		type: "error",
		error: result.error
	};
	const value$1 = result.value;
	if (typeof value$1 === "object" && value$1 != null && "help" in value$1 && "version" in value$1) {
		const parsedValue = value$1;
		const hasVersionOption = args.includes("--version");
		const hasVersionCommand = args.length > 0 && args[0] === "version";
		const hasHelpOption = args.includes("--help");
		const hasHelpCommand = args.length > 0 && args[0] === "help";
		const hasCompletionCommand = args.length > 0 && args[0] === "completion";
		if (hasVersionOption && hasHelpOption && !hasVersionCommand && !hasHelpCommand) {}
		if (hasVersionCommand && hasHelpOption && parsedValue.helpFlag) return {
			type: "help",
			commands: ["version"]
		};
		if (hasCompletionCommand && hasHelpOption && parsedValue.helpFlag) return {
			type: "help",
			commands: ["completion"]
		};
		if (parsedValue.help && (hasHelpOption || hasHelpCommand)) {
			let commandContext = [];
			if (Array.isArray(parsedValue.commands)) commandContext = parsedValue.commands;
			else if (typeof parsedValue.commands === "object" && parsedValue.commands != null && "length" in parsedValue.commands) commandContext = parsedValue.commands;
			return {
				type: "help",
				commands: commandContext
			};
		}
		if ((hasVersionOption || hasVersionCommand) && (parsedValue.version || parsedValue.versionFlag)) return { type: "version" };
		if (parsedValue.completion && parsedValue.completionData) return {
			type: "completion",
			shell: parsedValue.completionData.shell || "",
			args: parsedValue.completionData.args || []
		};
		return {
			type: "success",
			value: parsedValue.result ?? value$1
		};
	}
	return {
		type: "success",
		value: value$1
	};
}
/**
* Handles shell completion requests.
* @since 0.6.0
*/
function handleCompletion(completionArgs, programName, parser, completionParser, stdout, stderr, onCompletion, onError, availableShells, colors, maxWidth, completionMode, completionName) {
	const shellName = completionArgs[0] || "";
	const args = completionArgs.slice(1);
	const callOnError = (code) => {
		try {
			return onError(code);
		} catch {
			return onError();
		}
	};
	const callOnCompletion = (code) => {
		try {
			return onCompletion(code);
		} catch {
			return onCompletion();
		}
	};
	if (!shellName) {
		stderr("Error: Missing shell name for completion.\n");
		if (completionParser) {
			const doc = require_parser.getDocPage(completionParser, ["completion"]);
			if (doc) stderr(require_doc.formatDocPage(programName, doc, {
				colors,
				maxWidth
			}));
		}
		if (parser.$mode === "async") return Promise.resolve(callOnError(1));
		return callOnError(1);
	}
	const shell = availableShells[shellName];
	if (!shell) {
		const available = [];
		for (const name in availableShells) {
			if (available.length > 0) available.push(require_message.text(", "));
			available.push(require_message.value(name));
		}
		stderr(require_message.formatMessage(require_message.message`Error: Unsupported shell ${shellName}. Available shells: ${available}.`, {
			colors,
			quotes: !colors
		}));
		if (parser.$mode === "async") return Promise.resolve(callOnError(1));
		return callOnError(1);
	}
	if (args.length === 0) {
		const usePlural = completionName === "plural";
		const completionArg = completionMode === "option" ? usePlural ? "--completions" : "--completion" : usePlural ? "completions" : "completion";
		const script = shell.generateScript(programName, [completionArg, shellName]);
		stdout(script);
		if (parser.$mode === "async") return Promise.resolve(callOnCompletion(0));
		return callOnCompletion(0);
	}
	if (parser.$mode === "async") return (async () => {
		const suggestions$1 = await require_parser.suggestAsync(parser, args);
		for (const chunk of shell.encodeSuggestions(suggestions$1)) stdout(chunk);
		return callOnCompletion(0);
	})();
	const syncParser = parser;
	const suggestions = require_parser.suggest(syncParser, args);
	for (const chunk of shell.encodeSuggestions(suggestions)) stdout(chunk);
	return callOnCompletion(0);
}
function runParser(parserOrProgram, programNameOrArgs, argsOrOptions, optionsParam) {
	const isProgram = typeof programNameOrArgs !== "string";
	let parser;
	let programName;
	let args;
	let options;
	if (isProgram) {
		const program = parserOrProgram;
		parser = program.parser;
		programName = program.metadata.name;
		args = programNameOrArgs;
		options = argsOrOptions ?? {};
		options = {
			...options,
			brief: options.brief ?? program.metadata.brief,
			description: options.description ?? program.metadata.description,
			examples: options.examples ?? program.metadata.examples,
			author: options.author ?? program.metadata.author,
			bugs: options.bugs ?? program.metadata.bugs,
			footer: options.footer ?? program.metadata.footer
		};
	} else {
		parser = parserOrProgram;
		programName = programNameOrArgs;
		args = argsOrOptions;
		options = optionsParam ?? {};
	}
	const { colors, maxWidth, showDefault, showChoices, aboveError = "usage", onError = () => {
		throw new RunParserError("Failed to parse command line arguments.");
	}, stderr = console.error, stdout = console.log, brief, description, examples, author, bugs, footer } = options;
	const helpMode = options.help?.mode ?? "option";
	const onHelp = options.help?.onShow ?? (() => ({}));
	const helpGroup = options.help?.group;
	const versionMode = options.version?.mode ?? "option";
	const versionValue = options.version?.value ?? "";
	const onVersion = options.version?.onShow ?? (() => ({}));
	const versionGroup = options.version?.group;
	const completionMode = options.completion?.mode ?? "both";
	const completionName = options.completion?.name ?? "both";
	const completionHelpVisibility = options.completion?.helpVisibility ?? completionName;
	const onCompletion = options.completion?.onShow ?? (() => ({}));
	const completionGroup = options.completion?.group;
	const defaultShells = {
		bash: require_completion.bash,
		fish: require_completion.fish,
		nu: require_completion.nu,
		pwsh: require_completion.pwsh,
		zsh: require_completion.zsh
	};
	const availableShells = options.completion?.shells ? {
		...defaultShells,
		...options.completion.shells
	} : defaultShells;
	const help = options.help ? helpMode : "none";
	const version = options.version ? versionMode : "none";
	const completion = options.completion ? completionMode : "none";
	const helpParsers = help === "none" ? {
		helpCommand: null,
		helpOption: null
	} : createHelpParser(help);
	const versionParsers = version === "none" ? {
		versionCommand: null,
		versionOption: null
	} : createVersionParser(version);
	const completionParsers = completion === "none" ? {
		completionCommand: null,
		completionOption: null
	} : createCompletionParser(completion, programName, availableShells, completionName, completionHelpVisibility);
	if (options.completion) {
		const hasHelpOption = args.includes("--help");
		if ((completionMode === "command" || completionMode === "both") && args.length >= 1 && ((completionName === "singular" || completionName === "both" ? args[0] === "completion" : false) || (completionName === "plural" || completionName === "both" ? args[0] === "completions" : false)) && !hasHelpOption) return handleCompletion(args.slice(1), programName, parser, completionParsers.completionCommand, stdout, stderr, onCompletion, onError, availableShells, colors, maxWidth, completionMode, completionName);
		if (completionMode === "option" || completionMode === "both") for (let i = 0; i < args.length; i++) {
			const arg = args[i];
			const singularMatch = completionName === "singular" || completionName === "both" ? arg.startsWith("--completion=") : false;
			const pluralMatch = completionName === "plural" || completionName === "both" ? arg.startsWith("--completions=") : false;
			if (singularMatch || pluralMatch) {
				const shell = arg.slice(arg.indexOf("=") + 1);
				const completionArgs = args.slice(i + 1);
				return handleCompletion([shell, ...completionArgs], programName, parser, completionParsers.completionCommand, stdout, stderr, onCompletion, onError, availableShells, colors, maxWidth, completionMode, completionName);
			} else {
				const singularMatchExact = completionName === "singular" || completionName === "both" ? arg === "--completion" : false;
				const pluralMatchExact = completionName === "plural" || completionName === "both" ? arg === "--completions" : false;
				if (singularMatchExact || pluralMatchExact) {
					const shell = i + 1 < args.length ? args[i + 1] : "";
					const completionArgs = i + 1 < args.length ? args.slice(i + 2) : [];
					return handleCompletion([shell, ...completionArgs], programName, parser, completionParsers.completionCommand, stdout, stderr, onCompletion, onError, availableShells, colors, maxWidth, completionMode, completionName);
				}
			}
		}
	}
	const augmentedParser = help === "none" && version === "none" && completion === "none" ? parser : combineWithHelpVersion(parser, helpParsers, versionParsers, completionParsers, {
		helpGroup,
		versionGroup,
		completionGroup
	});
	const handleResult = (result) => {
		const classified = classifyResult(result, args);
		switch (classified.type) {
			case "success": return classified.value;
			case "version":
				stdout(versionValue);
				try {
					return onVersion(0);
				} catch {
					return onVersion();
				}
			case "completion": throw new RunParserError("Completion should be handled by early return");
			case "help": {
				let helpGeneratorParser;
				const helpAsCommand = help === "command" || help === "both";
				const versionAsCommand = version === "command" || version === "both";
				const completionAsCommand = completion === "command" || completion === "both";
				const helpAsOption = help === "option" || help === "both";
				const versionAsOption = version === "option" || version === "both";
				const completionAsOption = completion === "option" || completion === "both";
				const requestedCommand = classified.commands[0];
				if ((requestedCommand === "completion" || requestedCommand === "completions") && completionAsCommand && completionParsers.completionCommand) helpGeneratorParser = completionParsers.completionCommand;
				else if (requestedCommand === "help" && helpAsCommand && helpParsers.helpCommand) helpGeneratorParser = helpParsers.helpCommand;
				else if (requestedCommand === "version" && versionAsCommand && versionParsers.versionCommand) helpGeneratorParser = versionParsers.versionCommand;
				else {
					const commandParsers = [parser];
					const groupedMeta = {};
					const ungroupedMeta = [];
					const addMeta = (p, groupLabel) => {
						if (groupLabel) (groupedMeta[groupLabel] ??= []).push(p);
						else ungroupedMeta.push(p);
					};
					if (helpAsCommand && helpParsers.helpCommand) addMeta(helpParsers.helpCommand, helpGroup);
					if (versionAsCommand && versionParsers.versionCommand) addMeta(versionParsers.versionCommand, versionGroup);
					if (completionAsCommand && completionParsers.completionCommand) addMeta(completionParsers.completionCommand, completionGroup);
					commandParsers.push(...ungroupedMeta);
					for (const [label, parsers] of Object.entries(groupedMeta)) if (parsers.length === 1) commandParsers.push(require_constructs.group(label, parsers[0]));
					else commandParsers.push(require_constructs.group(label, require_constructs.longestMatch(...parsers)));
					if (helpAsOption && helpParsers.helpOption) commandParsers.push(helpParsers.helpOption);
					if (versionAsOption && versionParsers.versionOption) commandParsers.push(versionParsers.versionOption);
					if (completionAsOption && completionParsers.completionOption) commandParsers.push(completionParsers.completionOption);
					if (commandParsers.length === 1) helpGeneratorParser = commandParsers[0];
					else if (commandParsers.length === 2) helpGeneratorParser = require_constructs.longestMatch(commandParsers[0], commandParsers[1]);
					else helpGeneratorParser = require_constructs.longestMatch(...commandParsers);
				}
				const reportInvalidHelpCommand = (validationError) => {
					stderr(`Usage: ${indentLines(require_usage.formatUsage(programName, augmentedParser.usage, {
						colors,
						maxWidth: maxWidth == null ? void 0 : maxWidth - 7,
						expandCommands: true
					}), 7)}`);
					const errorMessage = require_message.formatMessage(validationError, {
						colors,
						quotes: !colors
					});
					stderr(`Error: ${errorMessage}`);
					return onError(1);
				};
				if (classified.commands.length > 0) {
					let validationContext = {
						buffer: [...classified.commands],
						optionsTerminated: false,
						state: helpGeneratorParser.initialState,
						usage: helpGeneratorParser.usage
					};
					const processStep = (stepResult) => {
						if (!stepResult.success) return stepResult.error;
						if (stepResult.consumed.length < 1) return require_message.message`Unexpected option or subcommand: ${require_message.optionName(validationContext.buffer[0])}.`;
						validationContext = {
							...validationContext,
							buffer: stepResult.next.buffer,
							optionsTerminated: stepResult.next.optionsTerminated,
							state: stepResult.next.state,
							usage: stepResult.next.usage ?? validationContext.usage
						};
						return validationContext.buffer.length > 0 ? "continue" : null;
					};
					let validationResult = "continue";
					while (validationResult === "continue") {
						const stepResult = helpGeneratorParser.parse(validationContext);
						if (stepResult instanceof Promise) {
							const asyncValidate = async (result$1) => {
								let res = processStep(result$1);
								while (res === "continue") {
									const next = helpGeneratorParser.parse(validationContext);
									const resolved = next instanceof Promise ? await next : next;
									res = processStep(resolved);
								}
								if (res != null) return reportInvalidHelpCommand(res);
								const docOrPromise$1 = require_parser.getDocPage(helpGeneratorParser, classified.commands);
								return docOrPromise$1 instanceof Promise ? docOrPromise$1.then(displayHelp) : displayHelp(docOrPromise$1);
							};
							return stepResult.then(asyncValidate);
						}
						validationResult = processStep(stepResult);
					}
					if (validationResult != null) return reportInvalidHelpCommand(validationResult);
				}
				const displayHelp = (doc) => {
					if (doc != null) {
						const isMetaCommandHelp = (completionName === "singular" || completionName === "both" ? requestedCommand === "completion" : false) || (completionName === "plural" || completionName === "both" ? requestedCommand === "completions" : false) || requestedCommand === "help" || requestedCommand === "version";
						const isSubcommandHelp = classified.commands.length > 0;
						const isTopLevel = !isSubcommandHelp;
						const shouldOverride = !isMetaCommandHelp && !isSubcommandHelp;
						const augmentedDoc = {
							...doc,
							brief: shouldOverride ? brief ?? doc.brief : doc.brief,
							description: shouldOverride ? description ?? doc.description : doc.description,
							examples: isTopLevel && !isMetaCommandHelp ? examples ?? doc.examples : void 0,
							author: isTopLevel && !isMetaCommandHelp ? author ?? doc.author : void 0,
							bugs: isTopLevel && !isMetaCommandHelp ? bugs ?? doc.bugs : void 0,
							footer: shouldOverride ? footer ?? doc.footer : doc.footer ?? footer
						};
						stdout(require_doc.formatDocPage(programName, augmentedDoc, {
							colors,
							maxWidth,
							showDefault,
							showChoices
						}));
					}
					try {
						return onHelp(0);
					} catch {
						return onHelp();
					}
				};
				const docOrPromise = require_parser.getDocPage(helpGeneratorParser, classified.commands);
				if (docOrPromise instanceof Promise) return docOrPromise.then(displayHelp);
				return displayHelp(docOrPromise);
			}
			case "error": {
				const displayError = (doc, currentAboveError) => {
					let effectiveAboveError = currentAboveError;
					if (effectiveAboveError === "help") if (doc == null) effectiveAboveError = "usage";
					else {
						const augmentedDoc = {
							...doc,
							brief: brief ?? doc.brief,
							description: description ?? doc.description,
							examples: examples ?? doc.examples,
							author: author ?? doc.author,
							bugs: bugs ?? doc.bugs,
							footer: footer ?? doc.footer
						};
						stderr(require_doc.formatDocPage(programName, augmentedDoc, {
							colors,
							maxWidth,
							showDefault,
							showChoices
						}));
					}
					if (effectiveAboveError === "usage") stderr(`Usage: ${indentLines(require_usage.formatUsage(programName, augmentedParser.usage, {
						colors,
						maxWidth: maxWidth == null ? void 0 : maxWidth - 7,
						expandCommands: true
					}), 7)}`);
					const errorMessage = require_message.formatMessage(classified.error, {
						colors,
						quotes: !colors
					});
					stderr(`Error: ${errorMessage}`);
					return onError(1);
				};
				if (aboveError === "help") {
					const parserForDoc = args.length < 1 ? augmentedParser : parser;
					const docOrPromise = require_parser.getDocPage(parserForDoc, args);
					if (docOrPromise instanceof Promise) return docOrPromise.then((doc) => displayError(doc, aboveError));
					return displayError(docOrPromise, aboveError);
				}
				return displayError(void 0, aboveError);
			}
			default: throw new RunParserError("Unexpected parse result type");
		}
	};
	if (parser.$mode === "async") return require_parser.parseAsync(augmentedParser, args).then(handleResult);
	else {
		const result = require_parser.parseSync(augmentedParser, args);
		return handleResult(result);
	}
}
/**
* Runs a synchronous command-line parser with the given options.
*
* This is a type-safe version of {@link runParser} that only accepts sync
* parsers. Use this when you know your parser is sync-only to get direct
* return values without Promise wrappers.
*
* @template TParser The sync parser type being executed.
* @template THelp The return type of the onHelp callback.
* @template TError The return type of the onError callback.
* @param parser The synchronous command-line parser to execute.
* @param programName The name of the program for help messages.
* @param args The command-line arguments to parse.
* @param options Configuration options for customizing behavior.
* @returns The parsed result if successful.
* @since 0.9.0
*/
function runParserSync(parser, programName, args, options) {
	return runParser(parser, programName, args, options);
}
/**
* Runs any command-line parser asynchronously with the given options.
*
* This function accepts parsers of any mode (sync or async) and always
* returns a Promise. Use this when working with parsers that may contain
* async value parsers.
*
* @template TParser The parser type being executed.
* @template THelp The return type of the onHelp callback.
* @template TError The return type of the onError callback.
* @param parser The command-line parser to execute.
* @param programName The name of the program for help messages.
* @param args The command-line arguments to parse.
* @param options Configuration options for customizing behavior.
* @returns A Promise of the parsed result if successful.
* @since 0.9.0
*/
function runParserAsync(parser, programName, args, options) {
	const result = runParser(parser, programName, args, options);
	return Promise.resolve(result);
}
/**
* An error class used to indicate that the command line arguments
* could not be parsed successfully.
*/
var RunParserError = class extends Error {
	constructor(message$1) {
		super(message$1);
		this.name = "RunParserError";
	}
};
function indentLines(text$1, indent) {
	return text$1.split("\n").join("\n" + " ".repeat(indent));
}
/**
* Checks if the arguments contain help, version, or completion requests
* that should be handled immediately without context processing.
*
* This enables early exit optimization: when users request help, version,
* or completion, we skip annotation collection and context processing
* entirely, delegating directly to runParser().
*
* @param args Command-line arguments to check.
* @param options Run options containing help/version/completion configuration.
* @returns `true` if early exit should be performed, `false` otherwise.
*/
function needsEarlyExit(args, options) {
	if (options.help) {
		const helpMode = options.help.mode ?? "option";
		if ((helpMode === "option" || helpMode === "both") && args.includes("--help")) return true;
		if ((helpMode === "command" || helpMode === "both") && args[0] === "help") return true;
	}
	if (options.version) {
		const versionMode = options.version.mode ?? "option";
		if ((versionMode === "option" || versionMode === "both") && args.includes("--version")) return true;
		if ((versionMode === "command" || versionMode === "both") && args[0] === "version") return true;
	}
	if (options.completion) {
		const completionMode = options.completion.mode ?? "both";
		const completionName = options.completion.name ?? "both";
		if (completionMode === "command" || completionMode === "both") {
			if ((completionName === "singular" || completionName === "both") && args[0] === "completion") return true;
			if ((completionName === "plural" || completionName === "both") && args[0] === "completions") return true;
		}
		if (completionMode === "option" || completionMode === "both") for (const arg of args) {
			if ((completionName === "singular" || completionName === "both") && (arg === "--completion" || arg.startsWith("--completion="))) return true;
			if ((completionName === "plural" || completionName === "both") && (arg === "--completions" || arg.startsWith("--completions="))) return true;
		}
	}
	return false;
}
/**
* Merges multiple annotation objects, with earlier contexts having priority.
*
* When the same symbol key exists in multiple annotations, the value from
* the earlier context (lower index in the array) takes precedence.
*
* @param annotationsList Array of annotations to merge.
* @returns Merged annotations object.
*/
function mergeAnnotations(annotationsList) {
	const result = {};
	for (let i = annotationsList.length - 1; i >= 0; i--) {
		const annotations = annotationsList[i];
		for (const key of Object.getOwnPropertySymbols(annotations)) result[key] = annotations[key];
	}
	return result;
}
/**
* Collects phase 1 annotations from all contexts and determines whether
* two-phase parsing is needed.
*
* @param contexts Source contexts to collect annotations from.
* @returns Promise with merged annotations and dynamic-context hint.
*/
async function collectPhase1Annotations(contexts) {
	const annotationsList = [];
	let hasDynamic = false;
	for (const context of contexts) {
		const result = context.getAnnotations();
		if (result instanceof Promise) {
			hasDynamic = true;
			annotationsList.push(await result);
		} else {
			if (Object.getOwnPropertySymbols(result).length === 0) hasDynamic = true;
			annotationsList.push(result);
		}
	}
	return {
		annotations: mergeAnnotations(annotationsList),
		hasDynamic
	};
}
/**
* Collects annotations from all contexts.
*
* @param contexts Source contexts to collect annotations from.
* @param parsed Optional parsed result from a previous parse pass.
* @returns Promise that resolves to merged annotations.
*/
async function collectAnnotations(contexts, parsed) {
	const annotationsList = [];
	for (const context of contexts) {
		const result = context.getAnnotations(parsed);
		annotationsList.push(result instanceof Promise ? await result : result);
	}
	return mergeAnnotations(annotationsList);
}
/**
* Collects phase 1 annotations from all contexts synchronously and determines
* whether two-phase parsing is needed.
*
* @param contexts Source contexts to collect annotations from.
* @returns Merged annotations with dynamic-context hint.
* @throws Error if any context returns a Promise.
*/
function collectPhase1AnnotationsSync(contexts) {
	const annotationsList = [];
	let hasDynamic = false;
	for (const context of contexts) {
		const result = context.getAnnotations();
		if (result instanceof Promise) throw new Error(`Context ${String(context.id)} returned a Promise in sync mode. Use runWith() or runWithAsync() for async contexts.`);
		if (Object.getOwnPropertySymbols(result).length === 0) hasDynamic = true;
		annotationsList.push(result);
	}
	return {
		annotations: mergeAnnotations(annotationsList),
		hasDynamic
	};
}
/**
* Collects annotations from all contexts synchronously.
*
* @param contexts Source contexts to collect annotations from.
* @param parsed Optional parsed result from a previous parse pass.
* @returns Merged annotations.
* @throws Error if any context returns a Promise.
*/
function collectAnnotationsSync(contexts, parsed) {
	const annotationsList = [];
	for (const context of contexts) {
		const result = context.getAnnotations(parsed);
		if (result instanceof Promise) throw new Error(`Context ${String(context.id)} returned a Promise in sync mode. Use runWith() or runWithAsync() for async contexts.`);
		annotationsList.push(result);
	}
	return mergeAnnotations(annotationsList);
}
/**
* Runs a parser with multiple source contexts.
*
* This function automatically handles static and dynamic contexts with proper
* priority. Earlier contexts in the array override later ones.
*
* The function uses a smart two-phase approach:
*
* 1. *Phase 1*: Collect annotations from all contexts (static contexts return
*    their data, dynamic contexts may return empty).
* 2. *First parse*: Parse with Phase 1 annotations.
* 3. *Phase 2*: Call `getAnnotations(parsed)` on all contexts with the first
*    parse result.
* 4. *Second parse*: Parse again with merged annotations from both phases.
*
* If all contexts are static (no dynamic contexts), the second parse is skipped
* for optimization.
*
* @template TParser The parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns Promise that resolves to the parsed result.
* @since 0.10.0
*
* @example
* ```typescript
* import { runWith } from "@optique/core/facade";
* import type { SourceContext } from "@optique/core/context";
*
* const envContext: SourceContext = {
*   id: Symbol.for("@myapp/env"),
*   getAnnotations() {
*     return { [Symbol.for("@myapp/env")]: process.env };
*   }
* };
*
* const result = await runWith(
*   parser,
*   "myapp",
*   [envContext],
*   { args: process.argv.slice(2) }
* );
* ```
*/
async function runWith(parser, programName, contexts, options) {
	const args = options?.args ?? [];
	if (needsEarlyExit(args, options)) {
		if (parser.$mode === "async") return runParser(parser, programName, args, options);
		return Promise.resolve(runParser(parser, programName, args, options));
	}
	if (contexts.length === 0) {
		if (parser.$mode === "async") return runParser(parser, programName, args, options);
		return Promise.resolve(runParser(parser, programName, args, options));
	}
	const { annotations: phase1Annotations, hasDynamic: needsTwoPhase } = await collectPhase1Annotations(contexts);
	if (!needsTwoPhase) {
		const augmentedParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		if (parser.$mode === "async") return runParser(augmentedParser, programName, args, options);
		return Promise.resolve(runParser(augmentedParser, programName, args, options));
	}
	const augmentedParser1 = injectAnnotationsIntoParser(parser, phase1Annotations);
	let firstPassResult;
	let firstPassFailed = false;
	try {
		if (parser.$mode === "async") firstPassResult = await require_parser.parseAsync(augmentedParser1, args);
		else firstPassResult = require_parser.parseSync(augmentedParser1, args);
		if (typeof firstPassResult === "object" && firstPassResult !== null && "success" in firstPassResult) {
			const result = firstPassResult;
			if (result.success) firstPassResult = result.value;
			else firstPassFailed = true;
		}
	} catch {
		firstPassFailed = true;
	}
	if (firstPassFailed) {
		const augmentedParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		if (parser.$mode === "async") return runParser(augmentedParser, programName, args, options);
		return Promise.resolve(runParser(augmentedParser, programName, args, options));
	}
	const phase2Annotations = await collectAnnotations(contexts, firstPassResult);
	const finalAnnotations = mergeAnnotations([phase1Annotations, phase2Annotations]);
	const augmentedParser2 = injectAnnotationsIntoParser(parser, finalAnnotations);
	if (parser.$mode === "async") return runParser(augmentedParser2, programName, args, options);
	return Promise.resolve(runParser(augmentedParser2, programName, args, options));
}
/**
* Runs a synchronous parser with multiple source contexts.
*
* This is the sync-only variant of {@link runWith}. All contexts must return
* annotations synchronously (not Promises).
*
* @template TParser The sync parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The synchronous parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns The parsed result.
* @throws Error if any context returns a Promise.
* @since 0.10.0
*/
function runWithSync(parser, programName, contexts, options) {
	const args = options?.args ?? [];
	if (needsEarlyExit(args, options)) return runParser(parser, programName, args, options);
	if (contexts.length === 0) return runParser(parser, programName, args, options);
	const { annotations: phase1Annotations, hasDynamic: needsTwoPhase } = collectPhase1AnnotationsSync(contexts);
	if (!needsTwoPhase) {
		const augmentedParser = injectAnnotationsIntoParser(parser, phase1Annotations);
		return runParser(augmentedParser, programName, args, options);
	}
	const augmentedParser1 = injectAnnotationsIntoParser(parser, phase1Annotations);
	let firstPassResult;
	try {
		const result = require_parser.parseSync(augmentedParser1, args);
		if (result.success) firstPassResult = result.value;
		else return runParser(augmentedParser1, programName, args, options);
	} catch {
		return runParser(augmentedParser1, programName, args, options);
	}
	const phase2Annotations = collectAnnotationsSync(contexts, firstPassResult);
	const finalAnnotations = mergeAnnotations([phase1Annotations, phase2Annotations]);
	const augmentedParser2 = injectAnnotationsIntoParser(parser, finalAnnotations);
	return runParser(augmentedParser2, programName, args, options);
}
/**
* Runs any parser asynchronously with multiple source contexts.
*
* This function accepts parsers of any mode (sync or async) and always
* returns a Promise. Use this when working with async contexts or parsers.
*
* @template TParser The parser type.
* @template THelp Return type when help is shown.
* @template TError Return type when an error occurs.
* @param parser The parser to execute.
* @param programName Name of the program for help/error output.
* @param contexts Source contexts to use (priority: earlier overrides later).
* @param options Run options including args, help, version, etc.
* @returns Promise that resolves to the parsed result.
* @since 0.10.0
*/
function runWithAsync(parser, programName, contexts, options) {
	return runWith(parser, programName, contexts, options);
}
/**
* Creates a new parser with annotations injected into its initial state.
*
* @param parser The original parser.
* @param annotations Annotations to inject.
* @returns A new parser with annotations in its initial state.
*/
function injectAnnotationsIntoParser(parser, annotations) {
	if (parser.initialState == null) return parser;
	const newInitialState = {
		...parser.initialState,
		[require_annotations.annotationKey]: annotations
	};
	return {
		...parser,
		initialState: newInitialState
	};
}

//#endregion
exports.RunParserError = RunParserError;
exports.runParser = runParser;
exports.runParserAsync = runParserAsync;
exports.runParserSync = runParserSync;
exports.runWith = runWith;
exports.runWithAsync = runWithAsync;
exports.runWithSync = runWithSync;
