// @optique/core/dist/index.cjs
const require_annotations = require('./annotations.cjs');
const require_message = require('./message.cjs');
const require_completion = require('./completion.cjs');
const require_dependency = require('./dependency.cjs');
const require_usage = require('./usage.cjs');
const require_constructs = require('./constructs.cjs');
const require_doc = require('./doc.cjs');
const require_modifiers = require('./modifiers.cjs');
const require_nonempty = require('./nonempty.cjs');
const require_valueparser = require('./valueparser.cjs');
const require_primitives = require('./primitives.cjs');
const require_parser = require('./parser.cjs');
const require_facade = require('./facade.cjs');

exports.DependencyRegistry = require_dependency.DependencyRegistry;
exports.DuplicateOptionError = require_constructs.DuplicateOptionError;
exports.RunParserError = require_facade.RunParserError;
exports.WithDefaultError = require_modifiers.WithDefaultError;
exports.annotationKey = require_annotations.annotationKey;
exports.argument = require_primitives.argument;
exports.bash = require_completion.bash;
exports.choice = require_valueparser.choice;
exports.cidr = require_valueparser.cidr;
exports.command = require_primitives.command;
exports.commandLine = require_message.commandLine;
exports.concat = require_constructs.concat;
exports.conditional = require_constructs.conditional;
exports.constant = require_primitives.constant;
exports.createDeferredParseState = require_dependency.createDeferredParseState;
exports.createDependencySourceState = require_dependency.createDependencySourceState;
exports.createPendingDependencySourceState = require_dependency.createPendingDependencySourceState;
exports.defaultValues = require_dependency.defaultValues;
exports.deferredParseMarker = require_dependency.deferredParseMarker;
exports.dependency = require_dependency.dependency;
exports.dependencyId = require_dependency.dependencyId;
exports.dependencyIds = require_dependency.dependencyIds;
exports.dependencySourceMarker = require_dependency.dependencySourceMarker;
exports.dependencySourceStateMarker = require_dependency.dependencySourceStateMarker;
exports.deriveFrom = require_dependency.deriveFrom;
exports.deriveFromAsync = require_dependency.deriveFromAsync;
exports.deriveFromSync = require_dependency.deriveFromSync;
exports.derivedValueParserMarker = require_dependency.derivedValueParserMarker;
exports.domain = require_valueparser.domain;
exports.email = require_valueparser.email;
exports.ensureNonEmptyString = require_nonempty.ensureNonEmptyString;
exports.envVar = require_message.envVar;
exports.extractArgumentMetavars = require_usage.extractArgumentMetavars;
exports.extractCommandNames = require_usage.extractCommandNames;
exports.extractOptionNames = require_usage.extractOptionNames;
exports.fish = require_completion.fish;
exports.flag = require_primitives.flag;
exports.float = require_valueparser.float;
exports.formatDependencyError = require_dependency.formatDependencyError;
exports.formatDocPage = require_doc.formatDocPage;
exports.formatMessage = require_message.formatMessage;
exports.formatUsage = require_usage.formatUsage;
exports.formatUsageTerm = require_usage.formatUsageTerm;
exports.getAnnotations = require_annotations.getAnnotations;
exports.getDefaultValuesFunction = require_dependency.getDefaultValuesFunction;
exports.getDependencyIds = require_dependency.getDependencyIds;
exports.getDocPage = require_parser.getDocPage;
exports.getDocPageAsync = require_parser.getDocPageAsync;
exports.getDocPageSync = require_parser.getDocPageSync;
exports.group = require_constructs.group;
exports.hostname = require_valueparser.hostname;
exports.integer = require_valueparser.integer;
exports.ip = require_valueparser.ip;
exports.ipv4 = require_valueparser.ipv4;
exports.ipv6 = require_valueparser.ipv6;
exports.isDeferredParseState = require_dependency.isDeferredParseState;
exports.isDependencySource = require_dependency.isDependencySource;
exports.isDependencySourceState = require_dependency.isDependencySourceState;
exports.isDerivedValueParser = require_dependency.isDerivedValueParser;
exports.isNonEmptyString = require_nonempty.isNonEmptyString;
exports.isPendingDependencySourceState = require_dependency.isPendingDependencySourceState;
exports.isValueParser = require_valueparser.isValueParser;
exports.isWrappedDependencySource = require_dependency.isWrappedDependencySource;
exports.lineBreak = require_message.lineBreak;
exports.link = require_message.link;
exports.locale = require_valueparser.locale;
exports.longestMatch = require_constructs.longestMatch;
exports.macAddress = require_valueparser.macAddress;
exports.map = require_modifiers.map;
exports.merge = require_constructs.merge;
exports.message = require_message.message;
exports.metavar = require_message.metavar;
exports.multiple = require_modifiers.multiple;
exports.nonEmpty = require_modifiers.nonEmpty;
exports.normalizeUsage = require_usage.normalizeUsage;
exports.nu = require_completion.nu;
exports.object = require_constructs.object;
exports.option = require_primitives.option;
exports.optionName = require_message.optionName;
exports.optionNames = require_message.optionNames;
exports.optional = require_modifiers.optional;
exports.or = require_constructs.or;
exports.parse = require_parser.parse;
exports.parseAsync = require_parser.parseAsync;
exports.parseSync = require_parser.parseSync;
exports.parseWithDependency = require_dependency.parseWithDependency;
exports.passThrough = require_primitives.passThrough;
exports.pendingDependencySourceStateMarker = require_dependency.pendingDependencySourceStateMarker;
exports.port = require_valueparser.port;
exports.portRange = require_valueparser.portRange;
exports.pwsh = require_completion.pwsh;
exports.runParser = require_facade.runParser;
exports.runParserAsync = require_facade.runParserAsync;
exports.runParserSync = require_facade.runParserSync;
exports.runWith = require_facade.runWith;
exports.runWithAsync = require_facade.runWithAsync;
exports.runWithSync = require_facade.runWithSync;
exports.socketAddress = require_valueparser.socketAddress;
exports.string = require_valueparser.string;
exports.suggest = require_parser.suggest;
exports.suggestAsync = require_parser.suggestAsync;
exports.suggestSync = require_parser.suggestSync;
exports.suggestWithDependency = require_dependency.suggestWithDependency;
exports.text = require_message.text;
exports.transformsDependencyValue = require_dependency.transformsDependencyValue;
exports.transformsDependencyValueMarker = require_dependency.transformsDependencyValueMarker;
exports.tuple = require_constructs.tuple;
exports.url = require_valueparser.url;
exports.uuid = require_valueparser.uuid;
exports.value = require_message.value;
exports.valueSet = require_message.valueSet;
exports.values = require_message.values;
exports.withDefault = require_modifiers.withDefault;
exports.wrappedDependencySourceMarker = require_dependency.wrappedDependencySourceMarker;
exports.zsh = require_completion.zsh;
