// @optique/core/dist/valueparser.js
import { message, text, valueSet } from "./message.js";
import { ensureNonEmptyString, isNonEmptyString } from "./nonempty.js";

//#region src/valueparser.ts
/**
* A predicate function that checks if an object is a {@link ValueParser}.
* @param object The object to check.
* @return `true` if the object is a {@link ValueParser}, `false` otherwise.
*/
function isValueParser(object) {
	return typeof object === "object" && object != null && "$mode" in object && (object.$mode === "sync" || object.$mode === "async") && "metavar" in object && typeof object.metavar === "string" && "parse" in object && typeof object.parse === "function" && "format" in object && typeof object.format === "function";
}
/**
* Implementation of the choice parser for both string and number types.
*/
function choice(choices, options = {}) {
	const metavar = options.metavar ?? "TYPE";
	ensureNonEmptyString(metavar);
	const isNumberChoice = choices.length > 0 && typeof choices[0] === "number";
	if (isNumberChoice) {
		const numberChoices = choices;
		const numberOptions = options;
		return {
			$mode: "sync",
			metavar,
			choices,
			parse(input) {
				const parsed = Number(input);
				if (Number.isNaN(parsed)) return {
					success: false,
					error: formatNumberChoiceError(input, numberChoices, numberOptions)
				};
				const index = numberChoices.indexOf(parsed);
				if (index < 0) return {
					success: false,
					error: formatNumberChoiceError(input, numberChoices, numberOptions)
				};
				return {
					success: true,
					value: numberChoices[index]
				};
			},
			format(value) {
				return String(value);
			},
			suggest(prefix) {
				return numberChoices.map((value) => String(value)).filter((valueStr) => valueStr.startsWith(prefix)).map((valueStr) => ({
					kind: "literal",
					text: valueStr
				}));
			}
		};
	}
	const stringChoices = choices;
	const stringOptions = options;
	const normalizedValues = stringOptions.caseInsensitive ? stringChoices.map((v) => v.toLowerCase()) : stringChoices;
	return {
		$mode: "sync",
		metavar,
		choices,
		parse(input) {
			const normalizedInput = stringOptions.caseInsensitive ? input.toLowerCase() : input;
			const index = normalizedValues.indexOf(normalizedInput);
			if (index < 0) return {
				success: false,
				error: formatStringChoiceError(input, stringChoices, stringOptions)
			};
			return {
				success: true,
				value: stringChoices[index]
			};
		},
		format(value) {
			return String(value);
		},
		suggest(prefix) {
			const normalizedPrefix = stringOptions.caseInsensitive ? prefix.toLowerCase() : prefix;
			return stringChoices.filter((value) => {
				const normalizedValue = stringOptions.caseInsensitive ? value.toLowerCase() : value;
				return normalizedValue.startsWith(normalizedPrefix);
			}).map((value) => ({
				kind: "literal",
				text: value
			}));
		}
	};
}
/**
* Formats error message for string choice parser.
*/
function formatStringChoiceError(input, choices, options) {
	if (options.errors?.invalidChoice) return typeof options.errors.invalidChoice === "function" ? options.errors.invalidChoice(input, choices) : options.errors.invalidChoice;
	return formatDefaultChoiceError(input, choices);
}
/**
* Formats error message for number choice parser.
*/
function formatNumberChoiceError(input, choices, options) {
	if (options.errors?.invalidChoice) return typeof options.errors.invalidChoice === "function" ? options.errors.invalidChoice(input, choices) : options.errors.invalidChoice;
	return formatDefaultChoiceError(input, choices);
}
/**
* Formats default error message for choice parser.
*/
function formatDefaultChoiceError(input, choices) {
	const choiceStrings = choices.map((c) => String(c));
	return message`Expected one of ${valueSet(choiceStrings, { locale: "en-US" })}, but got ${input}.`;
}
/**
* Creates a {@link ValueParser} for strings.
*
* This parser validates that the input is a string and optionally checks
* if it matches a specified regular expression pattern.
*
* **Security note**: When using the `pattern` option with user-defined or
* complex patterns, be aware of potential Regular Expression Denial of Service
* (ReDoS) attacks. See {@link StringOptions.pattern} for more details.
*
* @param options Configuration options for the string parser.
* @returns A {@link ValueParser} that parses strings according to the
*          specified options.
*/
function string(options = {}) {
	const metavar = options.metavar ?? "STRING";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (options.pattern != null && !options.pattern.test(input)) return {
				success: false,
				error: options.errors?.patternMismatch ? typeof options.errors.patternMismatch === "function" ? options.errors.patternMismatch(input, options.pattern) : options.errors.patternMismatch : message`Expected a string matching pattern ${text(options.pattern.source)}, but got ${input}.`
			};
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
/**
* Creates a ValueParser for parsing integer values from strings.
*
* This function provides two modes of operation:
*
* - Regular mode: Returns JavaScript numbers
*   (safe up to `Number.MAX_SAFE_INTEGER`)
* - `bigint` mode: Returns `bigint` values for arbitrarily large integers
*
* The parser validates that the input is a valid integer and optionally
* enforces minimum and maximum value constraints.
*
* @example
* ```typescript
* // Create a parser for regular integers
* const portParser = integer({ min: 1, max: 0xffff });
*
* // Create a parser for BigInt values
* const bigIntParser = integer({ type: "bigint", min: 0n });
*
* // Use the parser
* const result = portParser.parse("8080");
* if (result.success) {
*   console.log(`Port: ${result.value}`);
* } else {
*   console.error(result.error);
* }
* ```
*
* @param options Configuration options specifying the type and constraints.
* @returns A {@link ValueParser} that converts string input to the specified
*          integer type.
*/
function integer(options) {
	if (options?.type === "bigint") {
		const metavar$1 = options.metavar ?? "INTEGER";
		ensureNonEmptyString(metavar$1);
		return {
			$mode: "sync",
			metavar: metavar$1,
			parse(input) {
				let value;
				try {
					value = BigInt(input);
				} catch (e) {
					if (e instanceof SyntaxError) return {
						success: false,
						error: options.errors?.invalidInteger ? typeof options.errors.invalidInteger === "function" ? options.errors.invalidInteger(input) : options.errors.invalidInteger : message`Expected a valid integer, but got ${input}.`
					};
					throw e;
				}
				if (options.min != null && value < options.min) return {
					success: false,
					error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : message`Expected a value greater than or equal to ${text(options.min.toLocaleString("en"))}, but got ${input}.`
				};
				else if (options.max != null && value > options.max) return {
					success: false,
					error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : message`Expected a value less than or equal to ${text(options.max.toLocaleString("en"))}, but got ${input}.`
				};
				return {
					success: true,
					value
				};
			},
			format(value) {
				return value.toString();
			}
		};
	}
	const metavar = options?.metavar ?? "INTEGER";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (!input.match(/^-?\d+$/)) return {
				success: false,
				error: options?.errors?.invalidInteger ? typeof options.errors.invalidInteger === "function" ? options.errors.invalidInteger(input) : options.errors.invalidInteger : message`Expected a valid integer, but got ${input}.`
			};
			const value = Number.parseInt(input);
			if (options?.min != null && value < options.min) return {
				success: false,
				error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : message`Expected a value greater than or equal to ${text(options.min.toLocaleString("en"))}, but got ${input}.`
			};
			else if (options?.max != null && value > options.max) return {
				success: false,
				error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : message`Expected a value less than or equal to ${text(options.max.toLocaleString("en"))}, but got ${input}.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
/**
* Creates a {@link ValueParser} for floating-point numbers.
*
* This parser validates that the input is a valid floating-point number
* and optionally enforces minimum and maximum value constraints.
* @param options Configuration options for the float parser.
* @returns A {@link ValueParser} that parses strings into floating-point
*          numbers.
*/
function float(options = {}) {
	const floatRegex = /^[+-]?(?:(?:\d+\.?\d*)|(?:\d*\.\d+))(?:[eE][+-]?\d+)?$/;
	const metavar = options.metavar ?? "NUMBER";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			let value;
			const lowerInput = input.toLowerCase();
			if (lowerInput === "nan" && options.allowNaN) value = NaN;
			else if ((lowerInput === "infinity" || lowerInput === "+infinity") && options.allowInfinity) value = Infinity;
			else if (lowerInput === "-infinity" && options.allowInfinity) value = -Infinity;
			else if (floatRegex.test(input)) {
				value = Number(input);
				if (Number.isNaN(value)) return {
					success: false,
					error: options.errors?.invalidNumber ? typeof options.errors.invalidNumber === "function" ? options.errors.invalidNumber(input) : options.errors.invalidNumber : message`Expected a valid number, but got ${input}.`
				};
			} else return {
				success: false,
				error: options.errors?.invalidNumber ? typeof options.errors.invalidNumber === "function" ? options.errors.invalidNumber(input) : options.errors.invalidNumber : message`Expected a valid number, but got ${input}.`
			};
			if (options.min != null && value < options.min) return {
				success: false,
				error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, options.min) : options.errors.belowMinimum : message`Expected a value greater than or equal to ${text(options.min.toLocaleString("en"))}, but got ${input}.`
			};
			else if (options.max != null && value > options.max) return {
				success: false,
				error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, options.max) : options.errors.aboveMaximum : message`Expected a value less than or equal to ${text(options.max.toLocaleString("en"))}, but got ${input}.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
/**
* Creates a {@link ValueParser} for URL values.
*
* This parser validates that the input is a well-formed URL and optionally
* restricts the allowed protocols. The parsed result is a JavaScript `URL`
* object.
* @param options Configuration options for the URL parser.
* @returns A {@link ValueParser} that converts string input to `URL` objects.
*/
function url(options = {}) {
	const allowedProtocols = options.allowedProtocols?.map((p) => p.toLowerCase());
	const metavar = options.metavar ?? "URL";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (!URL.canParse(input)) return {
				success: false,
				error: options.errors?.invalidUrl ? typeof options.errors.invalidUrl === "function" ? options.errors.invalidUrl(input) : options.errors.invalidUrl : message`Invalid URL: ${input}.`
			};
			const url$1 = new URL(input);
			if (allowedProtocols != null && !allowedProtocols.includes(url$1.protocol)) return {
				success: false,
				error: options.errors?.disallowedProtocol ? typeof options.errors.disallowedProtocol === "function" ? options.errors.disallowedProtocol(url$1.protocol, options.allowedProtocols) : options.errors.disallowedProtocol : message`URL protocol ${url$1.protocol} is not allowed. Allowed protocols: ${allowedProtocols.join(", ")}.`
			};
			return {
				success: true,
				value: url$1
			};
		},
		format(value) {
			return value.href;
		},
		*suggest(prefix) {
			if (allowedProtocols && prefix.length > 0 && !prefix.includes("://")) for (const protocol of allowedProtocols) {
				const cleanProtocol = protocol.replace(/:+$/, "");
				if (cleanProtocol.startsWith(prefix.toLowerCase())) yield {
					kind: "literal",
					text: `${cleanProtocol}://`
				};
			}
		}
	};
}
/**
* Creates a {@link ValueParser} for locale values.
*
* This parser validates that the input is a well-formed locale identifier
* according to the Unicode Locale Identifier standard (BCP 47).
* The parsed result is a JavaScript `Intl.Locale` object.
* @param options Configuration options for the locale parser.
* @returns A {@link ValueParser} that converts string input to `Intl.Locale`
*          objects.
*/
function locale(options = {}) {
	const metavar = options.metavar ?? "LOCALE";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			let locale$1;
			try {
				locale$1 = new Intl.Locale(input);
			} catch (e) {
				if (e instanceof RangeError) return {
					success: false,
					error: options.errors?.invalidLocale ? typeof options.errors.invalidLocale === "function" ? options.errors.invalidLocale(input) : options.errors.invalidLocale : message`Invalid locale: ${input}.`
				};
				throw e;
			}
			return {
				success: true,
				value: locale$1
			};
		},
		format(value) {
			return value.baseName;
		},
		*suggest(prefix) {
			const commonLocales = [
				"en",
				"en-US",
				"en-GB",
				"en-CA",
				"en-AU",
				"en-NZ",
				"en-IE",
				"en-ZA",
				"en-IN",
				"es",
				"es-ES",
				"es-MX",
				"es-AR",
				"es-CL",
				"es-CO",
				"es-PE",
				"es-VE",
				"es-EC",
				"es-GT",
				"es-CU",
				"es-BO",
				"es-DO",
				"es-HN",
				"es-PY",
				"es-SV",
				"es-NI",
				"es-CR",
				"es-PA",
				"es-UY",
				"es-PR",
				"fr",
				"fr-FR",
				"fr-CA",
				"fr-BE",
				"fr-CH",
				"fr-LU",
				"fr-MC",
				"de",
				"de-DE",
				"de-AT",
				"de-CH",
				"de-BE",
				"de-LU",
				"de-LI",
				"it",
				"it-IT",
				"it-CH",
				"it-SM",
				"it-VA",
				"pt",
				"pt-BR",
				"pt-PT",
				"pt-AO",
				"pt-MZ",
				"pt-CV",
				"pt-GW",
				"pt-ST",
				"pt-TL",
				"ru",
				"ru-RU",
				"ru-BY",
				"ru-KZ",
				"ru-KG",
				"ru-MD",
				"uk",
				"uk-UA",
				"be",
				"be-BY",
				"bg",
				"bg-BG",
				"cs",
				"cs-CZ",
				"sk",
				"sk-SK",
				"sl",
				"sl-SI",
				"hr",
				"hr-HR",
				"sr",
				"sr-RS",
				"mk",
				"mk-MK",
				"ja",
				"ja-JP",
				"ko",
				"ko-KR",
				"zh",
				"zh-CN",
				"zh-TW",
				"zh-HK",
				"zh-SG",
				"zh-MO",
				"ar",
				"ar-SA",
				"ar-AE",
				"ar-BH",
				"ar-DZ",
				"ar-EG",
				"ar-IQ",
				"ar-JO",
				"ar-KW",
				"ar-LB",
				"ar-LY",
				"ar-MA",
				"ar-OM",
				"ar-QA",
				"ar-SY",
				"ar-TN",
				"ar-YE",
				"hi",
				"hi-IN",
				"bn",
				"bn-BD",
				"bn-IN",
				"ur",
				"ur-PK",
				"ta",
				"ta-IN",
				"te",
				"te-IN",
				"mr",
				"mr-IN",
				"gu",
				"gu-IN",
				"kn",
				"kn-IN",
				"ml",
				"ml-IN",
				"pa",
				"pa-IN",
				"tr",
				"tr-TR",
				"tr-CY",
				"pl",
				"pl-PL",
				"nl",
				"nl-NL",
				"nl-BE",
				"nl-SR",
				"sv",
				"sv-SE",
				"sv-FI",
				"da",
				"da-DK",
				"no",
				"no-NO",
				"nb",
				"nb-NO",
				"nn",
				"nn-NO",
				"fi",
				"fi-FI",
				"is",
				"is-IS",
				"el",
				"el-GR",
				"el-CY",
				"hu",
				"hu-HU",
				"ro",
				"ro-RO",
				"et",
				"et-EE",
				"lv",
				"lv-LV",
				"lt",
				"lt-LT",
				"mt",
				"mt-MT",
				"ga",
				"ga-IE",
				"cy",
				"cy-GB",
				"eu",
				"eu-ES",
				"ca",
				"ca-ES",
				"ca-AD",
				"th",
				"th-TH",
				"vi",
				"vi-VN",
				"id",
				"id-ID",
				"ms",
				"ms-MY",
				"ms-BN",
				"ms-SG",
				"tl",
				"tl-PH",
				"km",
				"km-KH",
				"my",
				"my-MM",
				"lo",
				"lo-LA",
				"si",
				"si-LK",
				"ne",
				"ne-NP",
				"sw",
				"sw-TZ",
				"sw-KE",
				"am",
				"am-ET",
				"ha",
				"ha-NG",
				"yo",
				"yo-NG",
				"ig",
				"ig-NG",
				"zu",
				"zu-ZA",
				"xh",
				"xh-ZA",
				"af",
				"af-ZA",
				"he",
				"he-IL",
				"fa",
				"fa-IR",
				"fa-AF"
			];
			for (const locale$1 of commonLocales) if (locale$1.toLowerCase().startsWith(prefix.toLowerCase())) yield {
				kind: "literal",
				text: locale$1
			};
		}
	};
}
/**
* Creates a {@link ValueParser} for UUID values.
*
* This parser validates that the input is a well-formed UUID string in the
* standard format: `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` where each `x`
* is a hexadecimal digit.  The parser can optionally restrict to specific
* UUID versions.
*
* @param options Configuration options for the UUID parser.
* @returns A {@link ValueParser} that converts string input to {@link Uuid}
*          strings.
*/
function uuid(options = {}) {
	const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
	const metavar = options.metavar ?? "UUID";
	ensureNonEmptyString(metavar);
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (!uuidRegex.test(input)) return {
				success: false,
				error: options.errors?.invalidUuid ? typeof options.errors.invalidUuid === "function" ? options.errors.invalidUuid(input) : options.errors.invalidUuid : message`Expected a valid UUID in format ${"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"}, but got ${input}.`
			};
			if (options.allowedVersions != null && options.allowedVersions.length > 0) {
				const versionChar = input.charAt(14);
				const version = parseInt(versionChar, 16);
				if (!options.allowedVersions.includes(version)) return {
					success: false,
					error: options.errors?.disallowedVersion ? typeof options.errors.disallowedVersion === "function" ? options.errors.disallowedVersion(version, options.allowedVersions) : options.errors.disallowedVersion : (() => {
						let expectedVersions = message``;
						let i = 0;
						for (const v of options.allowedVersions) {
							expectedVersions = i < 1 ? message`${expectedVersions}${v.toLocaleString("en")}` : i + 1 >= options.allowedVersions.length ? message`${expectedVersions}, or ${v.toLocaleString("en")}` : message`${expectedVersions}, ${v.toLocaleString("en")}`;
							i++;
						}
						return message`Expected UUID version ${expectedVersions}, but got version ${version.toLocaleString("en")}.`;
					})()
				};
			}
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
/**
* Creates a ValueParser for TCP/UDP port numbers.
*
* This parser validates that the input is a valid port number (1-65535 by default)
* and optionally enforces range constraints and well-known port restrictions.
*
* Port numbers are validated according to the following rules:
* - Must be a valid integer (no decimals, no scientific notation)
* - Must be within the range `[min, max]` (default `[1, 65535]`)
* - If `disallowWellKnown` is `true`, ports 1-1023 are rejected
*
* The parser provides two modes of operation:
* - Regular mode: Returns JavaScript numbers (safe for all port values)
* - `bigint` mode: Returns `bigint` values for consistency with other numeric types
*
* @example
* ```typescript
* // Basic port parser (1-65535)
* option("--port", port())
*
* // Custom range (non-privileged ports only)
* option("--port", port({ min: 1024, max: 65535 }))
*
* // Disallow well-known ports (reject 1-1023)
* option("--port", port({ disallowWellKnown: true }))
*
* // Development ports only
* option("--dev-port", port({ min: 3000, max: 9000 }))
*
* // Using bigint type
* option("--port", port({ type: "bigint" }))
* ```
*
* @param options Configuration options specifying the type and constraints.
* @returns A {@link ValueParser} that converts string input to port numbers.
* @since 0.10.0
*/
function port(options) {
	if (options?.type === "bigint") {
		const metavar$1 = options.metavar ?? "PORT";
		ensureNonEmptyString(metavar$1);
		const min$1 = options.min ?? 1n;
		const max$1 = options.max ?? 65535n;
		return {
			$mode: "sync",
			metavar: metavar$1,
			parse(input) {
				let value;
				try {
					value = BigInt(input);
				} catch (e) {
					if (e instanceof SyntaxError) return {
						success: false,
						error: options.errors?.invalidPort ? typeof options.errors.invalidPort === "function" ? options.errors.invalidPort(input) : options.errors.invalidPort : message`Expected a valid port number, but got ${input}.`
					};
					throw e;
				}
				if (value < min$1) return {
					success: false,
					error: options.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, min$1) : options.errors.belowMinimum : message`Expected a port number greater than or equal to ${text(min$1.toLocaleString("en"))}, but got ${input}.`
				};
				if (value > max$1) return {
					success: false,
					error: options.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, max$1) : options.errors.aboveMaximum : message`Expected a port number less than or equal to ${text(max$1.toLocaleString("en"))}, but got ${input}.`
				};
				if (options.disallowWellKnown && value >= 1n && value <= 1023n) return {
					success: false,
					error: options.errors?.wellKnownNotAllowed ? typeof options.errors.wellKnownNotAllowed === "function" ? options.errors.wellKnownNotAllowed(value) : options.errors.wellKnownNotAllowed : message`Port ${value.toLocaleString("en")} is a well-known port (1-1023) and may require elevated privileges.`
				};
				return {
					success: true,
					value
				};
			},
			format(value) {
				return value.toString();
			}
		};
	}
	const metavar = options?.metavar ?? "PORT";
	ensureNonEmptyString(metavar);
	const min = options?.min ?? 1;
	const max = options?.max ?? 65535;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (!input.match(/^-?\d+$/)) return {
				success: false,
				error: options?.errors?.invalidPort ? typeof options.errors.invalidPort === "function" ? options.errors.invalidPort(input) : options.errors.invalidPort : message`Expected a valid port number, but got ${input}.`
			};
			const value = Number.parseInt(input);
			if (value < min) return {
				success: false,
				error: options?.errors?.belowMinimum ? typeof options.errors.belowMinimum === "function" ? options.errors.belowMinimum(value, min) : options.errors.belowMinimum : message`Expected a port number greater than or equal to ${text(min.toLocaleString("en"))}, but got ${input}.`
			};
			if (value > max) return {
				success: false,
				error: options?.errors?.aboveMaximum ? typeof options.errors.aboveMaximum === "function" ? options.errors.aboveMaximum(value, max) : options.errors.aboveMaximum : message`Expected a port number less than or equal to ${text(max.toLocaleString("en"))}, but got ${input}.`
			};
			if (options?.disallowWellKnown && value >= 1 && value <= 1023) return {
				success: false,
				error: options.errors?.wellKnownNotAllowed ? typeof options.errors.wellKnownNotAllowed === "function" ? options.errors.wellKnownNotAllowed(value) : options.errors.wellKnownNotAllowed : message`Port ${value.toLocaleString("en")} is a well-known port (1-1023) and may require elevated privileges.`
			};
			return {
				success: true,
				value
			};
		},
		format(value) {
			return value.toString();
		}
	};
}
function isPrivateIp(octets) {
	if (octets[0] === 10) return true;
	if (octets[0] === 172 && octets[1] >= 16 && octets[1] <= 31) return true;
	if (octets[0] === 192 && octets[1] === 168) return true;
	return false;
}
function isLoopbackIp(octets) {
	return octets[0] === 127;
}
function isLinkLocalIp(octets) {
	return octets[0] === 169 && octets[1] === 254;
}
function isMulticastIp(octets) {
	return octets[0] >= 224 && octets[0] <= 239;
}
function isBroadcastIp(octets) {
	return octets.every((o) => o === 255);
}
function isZeroIp(octets) {
	return octets.every((o) => o === 0);
}
/**
* Creates a value parser for IPv4 addresses.
*
* This parser validates IPv4 addresses in dotted-decimal notation (e.g.,
* "192.168.1.1") and provides options to filter specific IP address types
* such as private, loopback, link-local, multicast, broadcast, and zero
* addresses.
*
* @param options The parser options.
* @returns A value parser for IPv4 addresses.
* @throws {TypeError} If the metavar is an empty string.
* @since 0.10.0
* @example
* ```typescript
* import { ipv4 } from "@optique/core/valueparser";
*
* // Basic IPv4 parser (allows all types)
* const address = ipv4();
*
* // Public IPs only (no private/loopback)
* const publicIp = ipv4({
*   allowPrivate: false,
*   allowLoopback: false
* });
*
* // Server binding (allow 0.0.0.0 and private IPs)
* const bindAddress = ipv4({
*   allowZero: true,
*   allowPrivate: true
* });
* ```
*/
function ipv4(options) {
	const metavar = options?.metavar ?? "IPV4";
	ensureNonEmptyString(metavar);
	const allowPrivate = options?.allowPrivate ?? true;
	const allowLoopback = options?.allowLoopback ?? true;
	const allowLinkLocal = options?.allowLinkLocal ?? true;
	const allowMulticast = options?.allowMulticast ?? true;
	const allowBroadcast = options?.allowBroadcast ?? true;
	const allowZero = options?.allowZero ?? true;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			const parts = input.split(".");
			if (parts.length !== 4) {
				const errorMsg = options?.errors?.invalidIpv4;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid IPv4 address, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			const octets = [];
			for (const part of parts) {
				if (part.length === 0) {
					const errorMsg = options?.errors?.invalidIpv4;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid IPv4 address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (part.trim() !== part) {
					const errorMsg = options?.errors?.invalidIpv4;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid IPv4 address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (part.length > 1 && part[0] === "0") {
					const errorMsg = options?.errors?.invalidIpv4;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid IPv4 address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				const octet = Number(part);
				if (!Number.isInteger(octet) || octet < 0 || octet > 255) {
					const errorMsg = options?.errors?.invalidIpv4;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid IPv4 address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				octets.push(octet);
			}
			const ipAddress = octets.join(".");
			if (!allowPrivate && isPrivateIp(octets)) {
				const errorMsg = options?.errors?.privateNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is a private IP address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLoopback && isLoopbackIp(octets)) {
				const errorMsg = options?.errors?.loopbackNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is a loopback address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLinkLocal && isLinkLocalIp(octets)) {
				const errorMsg = options?.errors?.linkLocalNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is a link-local address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowMulticast && isMulticastIp(octets)) {
				const errorMsg = options?.errors?.multicastNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is a multicast address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowBroadcast && isBroadcastIp(octets)) {
				const errorMsg = options?.errors?.broadcastNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is the broadcast address.`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowZero && isZeroIp(octets)) {
				const errorMsg = options?.errors?.zeroNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(ipAddress) : errorMsg ?? message`${ipAddress} is the zero address.`;
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: ipAddress
			};
		},
		format(value) {
			return value;
		}
	};
}
/**
* Creates a value parser for DNS hostnames.
*
* Validates hostnames according to RFC 1123:
* - Labels separated by dots
* - Each label: 1-63 characters
* - Labels can contain alphanumeric characters and hyphens
* - Labels cannot start or end with a hyphen
* - Total length ≤ 253 characters (default)
*
* @param options - Options for hostname validation.
* @returns A value parser for hostnames.
* @since 0.10.0
*
* @example
* ```typescript
* import { hostname } from "@optique/core/valueparser";
*
* // Basic hostname parser
* const host = hostname();
*
* // Allow wildcards for certificate validation
* const domain = hostname({ allowWildcard: true });
*
* // Reject localhost
* const remoteHost = hostname({ allowLocalhost: false });
* ```
*/
function hostname(options) {
	const metavar = options?.metavar ?? "HOST";
	ensureNonEmptyString(metavar);
	const allowWildcard = options?.allowWildcard ?? false;
	const allowUnderscore = options?.allowUnderscore ?? false;
	const allowLocalhost = options?.allowLocalhost ?? true;
	const maxLength = options?.maxLength ?? 253;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (input.length > maxLength) {
				const errorMsg = options?.errors?.tooLong;
				const msg = typeof errorMsg === "function" ? errorMsg(input, maxLength) : errorMsg ?? message`Hostname ${input} is too long (maximum ${text(maxLength.toString())} characters).`;
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLocalhost && input === "localhost") {
				const errorMsg = options?.errors?.localhostNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Hostname 'localhost' is not allowed.`;
				return {
					success: false,
					error: msg
				};
			}
			if (input.startsWith("*.")) {
				if (!allowWildcard) {
					const errorMsg = options?.errors?.wildcardNotAllowed;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Wildcard hostname ${input} is not allowed.`;
					return {
						success: false,
						error: msg
					};
				}
				const rest = input.slice(2);
				if (!rest || rest.includes("*")) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
			}
			if (!allowUnderscore && input.includes("_")) {
				const errorMsg = options?.errors?.underscoreNotAllowed;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Hostname ${input} contains underscore, which is not allowed.`;
				return {
					success: false,
					error: msg
				};
			}
			if (input.length === 0) {
				const errorMsg = options?.errors?.invalidHostname;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			const labels = input.split(".");
			for (const label of labels) {
				if (label.length === 0) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (label.length > 63) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (label === "*") continue;
				if (label.startsWith("-") || label.endsWith("-")) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				const allowedPattern = allowUnderscore ? /^[a-zA-Z0-9_-]+$/ : /^[a-zA-Z0-9-]+$/;
				if (!allowedPattern.test(label)) {
					const errorMsg = options?.errors?.invalidHostname;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid hostname, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
			}
			return {
				success: true,
				value: input
			};
		},
		format(value) {
			return value;
		}
	};
}
function email(options) {
	const metavar = options?.metavar ?? "EMAIL";
	ensureNonEmptyString(metavar);
	const allowMultiple = options?.allowMultiple ?? false;
	const allowDisplayName = options?.allowDisplayName ?? false;
	const lowercase = options?.lowercase ?? false;
	const allowedDomains = options?.allowedDomains;
	const atextRegex = /^[a-zA-Z0-9._+-]+$/;
	function validateEmail(input) {
		const trimmed = input.trim();
		let emailAddr = trimmed;
		if (allowDisplayName && trimmed.includes("<") && trimmed.endsWith(">")) {
			const match = trimmed.match(/<([^>]+)>$/);
			if (match) emailAddr = match[1].trim();
		}
		let atIndex = -1;
		if (emailAddr.startsWith("\"")) {
			const closingQuoteIndex = emailAddr.indexOf("\"", 1);
			if (closingQuoteIndex === -1) return null;
			atIndex = emailAddr.indexOf("@", closingQuoteIndex);
		} else atIndex = emailAddr.indexOf("@");
		if (atIndex === -1) return null;
		const lastAtIndex = emailAddr.lastIndexOf("@");
		if (atIndex !== lastAtIndex) return null;
		const localPart = emailAddr.substring(0, atIndex);
		const domain$1 = emailAddr.substring(atIndex + 1);
		if (!localPart || localPart.length === 0) return null;
		let isValidLocal = false;
		if (localPart.startsWith("\"") && localPart.endsWith("\"")) isValidLocal = localPart.length >= 2;
		else {
			const localParts = localPart.split(".");
			if (localPart.startsWith(".") || localPart.endsWith(".")) return null;
			isValidLocal = localParts.length > 0 && localParts.every((part) => part.length > 0 && atextRegex.test(part));
		}
		if (!isValidLocal) return null;
		if (!domain$1 || domain$1.length === 0) return null;
		if (!domain$1.includes(".")) return null;
		if (domain$1.startsWith(".") || domain$1.endsWith(".") || domain$1.startsWith("-") || domain$1.endsWith("-")) return null;
		const domainLabels = domain$1.split(".");
		for (const label of domainLabels) {
			if (label.length === 0 || label.length > 63) return null;
			if (label.startsWith("-") || label.endsWith("-")) return null;
			if (!/^[a-zA-Z0-9-]+$/.test(label)) return null;
		}
		const resultEmail = emailAddr;
		return lowercase ? resultEmail.toLowerCase() : resultEmail;
	}
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (allowMultiple) {
				const emails = input.split(",").map((e) => e.trim());
				const validatedEmails = [];
				for (const email$1 of emails) {
					const validated = validateEmail(email$1);
					if (validated === null) {
						const errorMsg = options?.errors?.invalidEmail;
						const msg = typeof errorMsg === "function" ? errorMsg(email$1) : errorMsg ?? message`Expected a valid email address, but got ${email$1}.`;
						return {
							success: false,
							error: msg
						};
					}
					if (allowedDomains && allowedDomains.length > 0) {
						const atIndex = validated.indexOf("@");
						const domain$1 = validated.substring(atIndex + 1).toLowerCase();
						const isAllowed = allowedDomains.some((allowed) => domain$1 === allowed.toLowerCase());
						if (!isAllowed) {
							const errorMsg = options?.errors?.domainNotAllowed;
							if (typeof errorMsg === "function") return {
								success: false,
								error: errorMsg(validated, allowedDomains)
							};
							const msg = errorMsg ?? [
								{
									type: "text",
									text: "Email domain "
								},
								{
									type: "value",
									value: domain$1
								},
								{
									type: "text",
									text: ` is not allowed. Allowed domains: ${allowedDomains.join(", ")}.`
								}
							];
							return {
								success: false,
								error: msg
							};
						}
					}
					validatedEmails.push(validated);
				}
				return {
					success: true,
					value: validatedEmails
				};
			} else {
				const validated = validateEmail(input);
				if (validated === null) {
					const errorMsg = options?.errors?.invalidEmail;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a valid email address, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				if (allowedDomains && allowedDomains.length > 0) {
					const atIndex = validated.indexOf("@");
					const domain$1 = validated.substring(atIndex + 1).toLowerCase();
					const isAllowed = allowedDomains.some((allowed) => domain$1 === allowed.toLowerCase());
					if (!isAllowed) {
						const errorMsg = options?.errors?.domainNotAllowed;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(validated, allowedDomains)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Email domain "
							},
							{
								type: "value",
								value: domain$1
							},
							{
								type: "text",
								text: ` is not allowed. Allowed domains: ${allowedDomains.join(", ")}.`
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				}
				return {
					success: true,
					value: validated
				};
			}
		},
		format(value) {
			if (Array.isArray(value)) return value.join(",");
			return value;
		}
	};
}
/**
* Creates a value parser for socket addresses in "host:port" format.
*
* Validates socket addresses with support for:
* - Hostnames and IPv4 addresses (IPv6 support coming in future versions)
* - Configurable host:port separator
* - Optional default port
* - Host type filtering (hostname only, IP only, or both)
* - Port range validation
*
* @param options - Options for socket address validation.
* @returns A value parser for socket addresses.
* @since 0.10.0
*
* @example
* ```typescript
* import { socketAddress } from "@optique/core/valueparser";
*
* // Basic socket address parser
* const endpoint = socketAddress({ requirePort: true });
*
* // With default port
* const server = socketAddress({ defaultPort: 80 });
*
* // IP addresses only
* const bind = socketAddress({
*   defaultPort: 8080,
*   host: { type: "ip" }
* });
* ```
*/
function socketAddress(options) {
	const metavar = options?.metavar ?? "HOST:PORT";
	ensureNonEmptyString(metavar);
	const separator = options?.separator ?? ":";
	const defaultPort = options?.defaultPort;
	const requirePort = options?.requirePort ?? false;
	const hostType = options?.host?.type ?? "both";
	const hostnameParser = hostname({
		...options?.host?.hostname,
		metavar: "HOST"
	});
	const ipParser = ipv4({
		...options?.host?.ip,
		metavar: "HOST"
	});
	const portParser = port({
		...options?.port,
		metavar: "PORT",
		type: "number"
	});
	function parseHost(hostInput) {
		if (hostType === "hostname") {
			const ipResult = ipParser.parse(hostInput);
			if (ipResult.success) return null;
			const result = hostnameParser.parse(hostInput);
			return result.success ? result.value : null;
		} else if (hostType === "ip") {
			const result = ipParser.parse(hostInput);
			return result.success ? result.value : null;
		} else {
			const ipResult = ipParser.parse(hostInput);
			if (ipResult.success) return ipResult.value;
			const hostnameResult = hostnameParser.parse(hostInput);
			return hostnameResult.success ? hostnameResult.value : null;
		}
	}
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			const trimmed = input.trim();
			const separatorIndex = trimmed.lastIndexOf(separator);
			let hostPart;
			let portPart;
			if (separatorIndex === -1) {
				hostPart = trimmed;
				portPart = void 0;
			} else {
				hostPart = trimmed.substring(0, separatorIndex);
				portPart = trimmed.substring(separatorIndex + separator.length);
			}
			const validatedHost = parseHost(hostPart);
			if (validatedHost === null) {
				const errorMsg = options?.errors?.invalidFormat;
				const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a socket address in format host${separator}port, but got ${input}.`;
				return {
					success: false,
					error: msg
				};
			}
			let validatedPort;
			if (portPart === void 0 || portPart === "") {
				if (requirePort) {
					const errorMsg = options?.errors?.missingPort;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Port number is required but was not specified.`;
					return {
						success: false,
						error: msg
					};
				}
				if (defaultPort !== void 0) validatedPort = defaultPort;
				else {
					const errorMsg = options?.errors?.missingPort;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Port number is required but was not specified.`;
					return {
						success: false,
						error: msg
					};
				}
			} else {
				const portResult = portParser.parse(portPart);
				if (!portResult.success) {
					const errorMsg = options?.errors?.invalidFormat;
					const msg = typeof errorMsg === "function" ? errorMsg(input) : errorMsg ?? message`Expected a socket address in format host${separator}port, but got ${input}.`;
					return {
						success: false,
						error: msg
					};
				}
				validatedPort = portResult.value;
			}
			return {
				success: true,
				value: {
					host: validatedHost,
					port: validatedPort
				}
			};
		},
		format(value) {
			return `${value.host}${separator}${value.port}`;
		}
	};
}
function portRange(options) {
	const metavar = options?.metavar ?? "PORT-PORT";
	ensureNonEmptyString(metavar);
	const separator = options?.separator ?? "-";
	const allowSingle = options?.allowSingle ?? false;
	const isBigInt = options?.type === "bigint";
	const portParser = isBigInt ? port({
		type: "bigint",
		min: options.min,
		max: options.max,
		disallowWellKnown: options.disallowWellKnown,
		errors: options.errors
	}) : port({
		type: "number",
		min: options?.min,
		max: options?.max,
		disallowWellKnown: options?.disallowWellKnown,
		errors: options?.errors
	});
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			const trimmed = input.trim();
			const separatorIndex = trimmed.indexOf(separator);
			if (separatorIndex === -1) {
				if (!allowSingle) {
					const errorMsg = options?.errors?.invalidFormat;
					if (typeof errorMsg === "function") return {
						success: false,
						error: errorMsg(input)
					};
					const msg = errorMsg ?? [
						{
							type: "text",
							text: `Expected a port range in format start${separator}end, but got `
						},
						{
							type: "value",
							value: input
						},
						{
							type: "text",
							text: "."
						}
					];
					return {
						success: false,
						error: msg
					};
				}
				const portResult = portParser.parse(trimmed);
				if (!portResult.success) return portResult;
				const portValue = portResult.value;
				return {
					success: true,
					value: isBigInt ? {
						start: portValue,
						end: portValue
					} : {
						start: portValue,
						end: portValue
					}
				};
			}
			const startPart = trimmed.substring(0, separatorIndex);
			const endPart = trimmed.substring(separatorIndex + separator.length);
			const startResult = portParser.parse(startPart);
			if (!startResult.success) return startResult;
			const endResult = portParser.parse(endPart);
			if (!endResult.success) return endResult;
			const startValue = startResult.value;
			const endValue = endResult.value;
			if (isBigInt) {
				const start = startValue;
				const end = endValue;
				if (start > end) {
					const errorMsg = options.errors?.invalidRange;
					const msg = typeof errorMsg === "function" ? errorMsg(start, end) : errorMsg ?? message`Start port ${startPart} must be less than or equal to end port ${endPart}.`;
					return {
						success: false,
						error: msg
					};
				}
				return {
					success: true,
					value: {
						start,
						end
					}
				};
			} else {
				const start = startValue;
				const end = endValue;
				if (start > end) {
					const errorMsg = options?.errors?.invalidRange;
					const msg = typeof errorMsg === "function" ? errorMsg(start, end) : errorMsg ?? message`Start port ${startPart} must be less than or equal to end port ${endPart}.`;
					return {
						success: false,
						error: msg
					};
				}
				return {
					success: true,
					value: {
						start,
						end
					}
				};
			}
		},
		format(value) {
			return `${value.start}${separator}${value.end}`;
		}
	};
}
/**
* Creates a value parser for MAC (Media Access Control) addresses.
*
* Validates MAC-48 addresses (6 octets, 12 hex digits) in various formats:
* - Colon-separated: `00:1A:2B:3C:4D:5E`
* - Hyphen-separated: `00-1A-2B-3C-4D-5E`
* - Dot-separated (Cisco): `001A.2B3C.4D5E`
* - No separator: `001A2B3C4D5E`
*
* Returns the MAC address as a formatted string according to `case` and
* `outputSeparator` options.
*
* @param options Configuration options for the MAC address parser.
* @returns A parser that validates MAC addresses and returns formatted strings.
* @since 0.10.0
*
* @example
* ```typescript
* import { macAddress } from "@optique/core/valueparser";
*
* // Accept any format
* const mac = macAddress();
*
* // Normalize to uppercase colon-separated
* const normalizedMac = macAddress({
*   outputSeparator: ":",
*   case: "upper"
* });
* ```
*/
function macAddress(options) {
	const separator = options?.separator ?? "any";
	const caseOption = options?.case ?? "preserve";
	const outputSeparator = options?.outputSeparator;
	const metavar = options?.metavar ?? "MAC";
	const colonRegex = /^([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2}):([0-9a-fA-F]{1,2})$/;
	const hyphenRegex = /^([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})-([0-9a-fA-F]{1,2})$/;
	const dotRegex = /^([0-9a-fA-F]{4})\.([0-9a-fA-F]{4})\.([0-9a-fA-F]{4})$/;
	const noneRegex = /^([0-9a-fA-F]{12})$/;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			let octets = [];
			let inputSeparator;
			if (separator === ":" || separator === "any") {
				const match = colonRegex.exec(input);
				if (match) {
					octets = match.slice(1, 7);
					inputSeparator = ":";
				}
			}
			if (octets.length === 0 && (separator === "-" || separator === "any")) {
				const match = hyphenRegex.exec(input);
				if (match) {
					octets = match.slice(1, 7);
					inputSeparator = "-";
				}
			}
			if (octets.length === 0 && (separator === "." || separator === "any")) {
				const match = dotRegex.exec(input);
				if (match) {
					const groups = match.slice(1, 4);
					octets = groups.flatMap((group) => [group.slice(0, 2), group.slice(2, 4)]);
					inputSeparator = ".";
				}
			}
			if (octets.length === 0 && (separator === "none" || separator === "any")) {
				const match = noneRegex.exec(input);
				if (match) {
					const hex = match[1];
					octets = [];
					for (let i = 0; i < 12; i += 2) octets.push(hex.slice(i, i + 2));
					inputSeparator = "none";
				}
			}
			if (octets.length === 0) {
				const errorMsg = options?.errors?.invalidMacAddress;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid MAC address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			let formattedOctets = octets;
			if (caseOption === "upper") formattedOctets = octets.map((octet) => octet.toUpperCase());
			else if (caseOption === "lower") formattedOctets = octets.map((octet) => octet.toLowerCase());
			const finalSeparator = outputSeparator ?? inputSeparator ?? ":";
			let result;
			if (finalSeparator === ":") result = formattedOctets.join(":");
			else if (finalSeparator === "-") result = formattedOctets.join("-");
			else if (finalSeparator === ".") result = [
				formattedOctets[0] + formattedOctets[1],
				formattedOctets[2] + formattedOctets[3],
				formattedOctets[4] + formattedOctets[5]
			].join(".");
			else result = formattedOctets.join("");
			return {
				success: true,
				value: result
			};
		},
		format() {
			return metavar;
		}
	};
}
/**
* Creates a value parser for domain names.
*
* Validates domain names according to RFC 1035 with configurable options for
* subdomain filtering, TLD restrictions, minimum label requirements, and case
* normalization.
*
* @param options Parser options for domain validation.
* @returns A parser that accepts valid domain names as strings.
*
* @example
* ``` typescript
* import { option } from "@optique/core/primitives";
* import { domain } from "@optique/core/valueparser";
*
* // Accept any valid domain
* option("--domain", domain())
*
* // Root domains only (no subdomains)
* option("--root", domain({ allowSubdomains: false }))
*
* // Restrict to specific TLDs
* option("--domain", domain({ allowedTLDs: ["com", "org", "net"] }))
*
* // Normalize to lowercase
* option("--domain", domain({ lowercase: true }))
* ```
*
* @since 0.10.0
*/
function domain(options) {
	const metavar = options?.metavar ?? "DOMAIN";
	const allowSubdomains = options?.allowSubdomains ?? true;
	const allowedTLDs = options?.allowedTLDs;
	const minLabels = options?.minLabels ?? 2;
	const lowercase = options?.lowercase ?? false;
	const errors = options?.errors;
	const labelRegex = /^[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$/;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			if (input.length === 0 || input.startsWith(".") || input.endsWith(".")) {
				const errorMsg = errors?.invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (input.includes("..")) {
				const errorMsg = errors?.invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const labels = input.split(".");
			for (const label of labels) if (!labelRegex.test(label)) {
				const errorMsg = errors?.invalidDomain;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid domain name, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (labels.length < minLabels) {
				const errorMsg = errors?.tooFewLabels;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input, minLabels)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Domain "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: ` must have at least ${minLabels} labels.`
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowSubdomains && labels.length > 2) {
				const errorMsg = errors?.subdomainsNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Subdomains are not allowed, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (allowedTLDs !== void 0) {
				const tld = labels[labels.length - 1];
				const tldLower = tld.toLowerCase();
				const allowedTLDsLower = allowedTLDs.map((t) => t.toLowerCase());
				if (!allowedTLDsLower.includes(tldLower)) {
					const errorMsg = errors?.tldNotAllowed;
					if (typeof errorMsg === "function") return {
						success: false,
						error: errorMsg(tld, allowedTLDs)
					};
					const msg = errorMsg ?? [
						{
							type: "text",
							text: "Top-level domain "
						},
						{
							type: "value",
							value: tld
						},
						{
							type: "text",
							text: ` is not allowed. Allowed TLDs: ${allowedTLDs.join(", ")}.`
						}
					];
					return {
						success: false,
						error: msg
					};
				}
			}
			const result = lowercase ? input.toLowerCase() : input;
			return {
				success: true,
				value: result
			};
		},
		format() {
			return metavar;
		}
	};
}
/**
* Creates a value parser for IPv6 addresses.
*
* Validates and normalizes IPv6 addresses to canonical form (lowercase,
* compressed using `::` notation where appropriate).
*
* @param options Configuration options for IPv6 validation.
* @returns A value parser that validates IPv6 addresses.
*
* @example
* ```typescript
* // Basic IPv6 parser
* option("--ipv6", ipv6())
*
* // Global unicast only (no link-local, no unique local)
* option("--public-ipv6", ipv6({
*   allowLinkLocal: false,
*   allowUniqueLocal: false
* }))
* ```
*
* @since 0.10.0
*/
function ipv6(options) {
	const allowLoopback = options?.allowLoopback ?? true;
	const allowLinkLocal = options?.allowLinkLocal ?? true;
	const allowUniqueLocal = options?.allowUniqueLocal ?? true;
	const allowMulticast = options?.allowMulticast ?? true;
	const allowZero = options?.allowZero ?? true;
	const errors = options?.errors;
	const metavar = options?.metavar ?? "IPV6";
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			const normalized = parseAndNormalizeIpv6(input);
			if (normalized === null) {
				const errorMsg = errors?.invalidIpv6;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid IPv6 address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowZero && normalized === "::") {
				const errorMsg = errors?.zeroNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is the zero address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowLoopback && normalized === "::1") {
				const errorMsg = errors?.loopbackNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a loopback address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			const groups = expandIpv6(normalized);
			if (groups === null) {
				const errorMsg = errors?.invalidIpv6;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid IPv6 address, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const firstGroup = parseInt(groups[0], 16);
			if (!allowLinkLocal && (firstGroup & 65472) === 65152) {
				const errorMsg = errors?.linkLocalNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a link-local address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowUniqueLocal && (firstGroup & 65024) === 64512) {
				const errorMsg = errors?.uniqueLocalNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a unique local address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			if (!allowMulticast && (firstGroup & 65280) === 65280) {
				const errorMsg = errors?.multicastNotAllowed;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(normalized)
				};
				const msg = errorMsg ?? [{
					type: "value",
					value: normalized
				}, {
					type: "text",
					text: " is a multicast address."
				}];
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: normalized
			};
		},
		format() {
			return metavar;
		}
	};
}
/**
* Parses and normalizes an IPv6 address to canonical form.
* Returns null if the input is not a valid IPv6 address.
*/
function parseAndNormalizeIpv6(input) {
	if (input.length === 0) return null;
	const ipv4MappedMatch = input.match(/^(.+):(\d+\.\d+\.\d+\.\d+)$/);
	if (ipv4MappedMatch) {
		const ipv6Part = ipv4MappedMatch[1];
		const ipv4Part = ipv4MappedMatch[2];
		const ipv4Octets = ipv4Part.split(".");
		if (ipv4Octets.length !== 4) return null;
		const octets = ipv4Octets.map((o) => parseInt(o, 10));
		if (octets.some((o) => isNaN(o) || o < 0 || o > 255)) return null;
		const group1 = octets[0] << 8 | octets[1];
		const group2 = octets[2] << 8 | octets[3];
		const fullAddress = `${ipv6Part}:${group1.toString(16)}:${group2.toString(16)}`;
		return parseAndNormalizeIpv6(fullAddress);
	}
	const compressionCount = (input.match(/::/g) || []).length;
	if (compressionCount > 1) return null;
	let groups;
	if (input.includes("::")) {
		const parts = input.split("::");
		if (parts.length > 2) return null;
		const leftGroups = parts[0] ? parts[0].split(":") : [];
		const rightGroups = parts[1] ? parts[1].split(":") : [];
		const totalGroups = leftGroups.length + rightGroups.length;
		if (totalGroups > 8) return null;
		const zeroCount = 8 - totalGroups;
		const zeros = Array(zeroCount).fill("0");
		groups = [
			...leftGroups,
			...zeros,
			...rightGroups
		];
	} else {
		groups = input.split(":");
		if (groups.length !== 8) return null;
	}
	for (let i = 0; i < groups.length; i++) {
		const group = groups[i];
		if (group.length === 0 || group.length > 4) return null;
		if (!/^[0-9a-fA-F]+$/.test(group)) return null;
		groups[i] = parseInt(group, 16).toString(16);
	}
	return compressIpv6(groups);
}
/**
* Expands a compressed IPv6 address to 8 groups of 4 hex digits.
* Returns null if the input is invalid.
*/
function expandIpv6(input) {
	if (input.includes("::")) {
		const parts = input.split("::");
		if (parts.length > 2) return null;
		const leftGroups = parts[0] ? parts[0].split(":").filter((g) => g) : [];
		const rightGroups = parts[1] ? parts[1].split(":").filter((g) => g) : [];
		const totalGroups = leftGroups.length + rightGroups.length;
		if (totalGroups > 8) return null;
		const zeroCount = 8 - totalGroups;
		const zeros = Array(zeroCount).fill("0");
		const groups = [
			...leftGroups,
			...zeros,
			...rightGroups
		];
		return groups.map((g) => g.padStart(4, "0"));
	} else {
		const groups = input.split(":");
		if (groups.length !== 8) return null;
		return groups.map((g) => g.padStart(4, "0"));
	}
}
/**
* Compresses an IPv6 address by replacing the longest sequence of zeros with ::.
*/
function compressIpv6(groups) {
	let longestStart = -1;
	let longestLength = 0;
	let currentStart = -1;
	let currentLength = 0;
	for (let i = 0; i < groups.length; i++) if (groups[i] === "0") if (currentStart === -1) {
		currentStart = i;
		currentLength = 1;
	} else currentLength++;
	else {
		if (currentLength > longestLength) {
			longestStart = currentStart;
			longestLength = currentLength;
		}
		currentStart = -1;
		currentLength = 0;
	}
	if (currentLength > longestLength) {
		longestStart = currentStart;
		longestLength = currentLength;
	}
	if (longestLength < 2) return groups.join(":");
	const before = groups.slice(0, longestStart);
	const after = groups.slice(longestStart + longestLength);
	if (before.length === 0 && after.length === 0) return "::";
	else if (before.length === 0) return "::" + after.join(":");
	else if (after.length === 0) return before.join(":") + "::";
	else return before.join(":") + "::" + after.join(":");
}
/**
* Creates a value parser that accepts both IPv4 and IPv6 addresses.
*
* By default, accepts both IPv4 and IPv6 addresses. Use the `version` option
* to restrict to a specific IP version.
*
* @param options Configuration options for IP validation.
* @returns A value parser that validates IP addresses.
*
* @example
* ```typescript
* // Accept both IPv4 and IPv6
* option("--ip", ip())
*
* // IPv4 only
* option("--ipv4", ip({ version: 4 }))
*
* // Public IPs only (both versions)
* option("--public-ip", ip({
*   ipv4: { allowPrivate: false, allowLoopback: false },
*   ipv6: { allowLinkLocal: false, allowUniqueLocal: false }
* }))
* ```
*
* @since 0.10.0
*/
function ip(options) {
	const version = options?.version ?? "both";
	const metavar = options?.metavar ?? "IP";
	const errors = options?.errors;
	const ipv4Parser = version === 4 || version === "both" ? ipv4({
		...options?.ipv4,
		errors: {
			invalidIpv4: errors?.invalidIP,
			privateNotAllowed: errors?.privateNotAllowed,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			broadcastNotAllowed: errors?.broadcastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed
		}
	}) : null;
	const ipv6Parser = version === 6 || version === "both" ? ipv6({
		...options?.ipv6,
		errors: {
			invalidIpv6: errors?.invalidIP,
			loopbackNotAllowed: errors?.loopbackNotAllowed,
			linkLocalNotAllowed: errors?.linkLocalNotAllowed,
			uniqueLocalNotAllowed: errors?.uniqueLocalNotAllowed,
			multicastNotAllowed: errors?.multicastNotAllowed,
			zeroNotAllowed: errors?.zeroNotAllowed
		}
	}) : null;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			let ipv4Error = null;
			let ipv6Error = null;
			if (ipv4Parser !== null) {
				const result = ipv4Parser.parse(input);
				if (result.success) return result;
				ipv4Error = result;
				if (version === 4) return result;
			}
			if (ipv6Parser !== null) {
				const result = ipv6Parser.parse(input);
				if (result.success) return result;
				ipv6Error = result;
				if (version === 6) return result;
			}
			if (ipv4Error !== null && !ipv4Error.success) {
				const isGeneric = ipv4Error.error.some((term) => term.type === "text" && term.text.includes("Expected"));
				if (!isGeneric) return ipv4Error;
			}
			if (ipv6Error !== null && !ipv6Error.success) {
				const isGeneric = ipv6Error.error.some((term) => term.type === "text" && term.text.includes("Expected"));
				if (!isGeneric) return ipv6Error;
			}
			const errorMsg = errors?.invalidIP;
			if (typeof errorMsg === "function") return {
				success: false,
				error: errorMsg(input)
			};
			const msg = errorMsg ?? [
				{
					type: "text",
					text: "Expected a valid IP address, but got "
				},
				{
					type: "value",
					value: input
				},
				{
					type: "text",
					text: "."
				}
			];
			return {
				success: false,
				error: msg
			};
		},
		format() {
			return metavar;
		}
	};
}
/**
* Creates a value parser for CIDR notation (IP address with prefix length).
*
* Parses and validates CIDR notation like `192.168.0.0/24` or `2001:db8::/32`.
* Returns a structured object with the normalized IP address, prefix length,
* and IP version.
*
* @param options Configuration options for CIDR validation.
* @returns A value parser that validates CIDR notation.
*
* @example
* ```typescript
* // Accept both IPv4 and IPv6 CIDR
* option("--network", cidr())
*
* // IPv4 CIDR only with prefix constraints
* option("--subnet", cidr({
*   version: 4,
*   minPrefix: 16,
*   maxPrefix: 24
* }))
* ```
*
* @since 0.10.0
*/
function cidr(options) {
	const version = options?.version ?? "both";
	const minPrefix = options?.minPrefix;
	const maxPrefix = options?.maxPrefix;
	const errors = options?.errors;
	const metavar = options?.metavar ?? "CIDR";
	const ipv4Parser = version === 4 || version === "both" ? ipv4(options?.ipv4) : null;
	const ipv6Parser = version === 6 || version === "both" ? ipv6(options?.ipv6) : null;
	return {
		$mode: "sync",
		metavar,
		parse(input) {
			const slashIndex = input.lastIndexOf("/");
			if (slashIndex === -1) {
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			const ipPart = input.slice(0, slashIndex);
			const prefixPart = input.slice(slashIndex + 1);
			const prefix = parseInt(prefixPart, 10);
			if (!Number.isInteger(prefix) || prefixPart !== prefix.toString() || prefix < 0) {
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			let ipVersion = null;
			let normalizedIp = null;
			if (ipv4Parser !== null) {
				const result = ipv4Parser.parse(ipPart);
				if (result.success) {
					ipVersion = 4;
					normalizedIp = result.value;
					if (prefix > 32) {
						const errorMsg = errors?.invalidPrefix;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(prefix, 4)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Expected a prefix length between 0 and "
							},
							{
								type: "text",
								text: "32"
							},
							{
								type: "text",
								text: " for IPv4, but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				}
			}
			if (ipVersion === null && ipv6Parser !== null) {
				const result = ipv6Parser.parse(ipPart);
				if (result.success) {
					ipVersion = 6;
					normalizedIp = result.value;
					if (prefix > 128) {
						const errorMsg = errors?.invalidPrefix;
						if (typeof errorMsg === "function") return {
							success: false,
							error: errorMsg(prefix, 6)
						};
						const msg = errorMsg ?? [
							{
								type: "text",
								text: "Expected a prefix length between 0 and "
							},
							{
								type: "text",
								text: "128"
							},
							{
								type: "text",
								text: " for IPv6, but got "
							},
							{
								type: "text",
								text: prefix.toString()
							},
							{
								type: "text",
								text: "."
							}
						];
						return {
							success: false,
							error: msg
						};
					}
				}
			}
			if (ipVersion === null || normalizedIp === null) {
				const errorMsg = errors?.invalidCidr;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(input)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a valid CIDR notation, but got "
					},
					{
						type: "value",
						value: input
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (minPrefix !== void 0 && prefix < minPrefix) {
				const errorMsg = errors?.prefixBelowMinimum;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(prefix, minPrefix)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a prefix length greater than or equal to "
					},
					{
						type: "text",
						text: minPrefix.toString()
					},
					{
						type: "text",
						text: ", but got "
					},
					{
						type: "text",
						text: prefix.toString()
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			if (maxPrefix !== void 0 && prefix > maxPrefix) {
				const errorMsg = errors?.prefixAboveMaximum;
				if (typeof errorMsg === "function") return {
					success: false,
					error: errorMsg(prefix, maxPrefix)
				};
				const msg = errorMsg ?? [
					{
						type: "text",
						text: "Expected a prefix length less than or equal to "
					},
					{
						type: "text",
						text: maxPrefix.toString()
					},
					{
						type: "text",
						text: ", but got "
					},
					{
						type: "text",
						text: prefix.toString()
					},
					{
						type: "text",
						text: "."
					}
				];
				return {
					success: false,
					error: msg
				};
			}
			return {
				success: true,
				value: {
					address: normalizedIp,
					prefix,
					version: ipVersion
				}
			};
		},
		format() {
			return metavar;
		}
	};
}

//#endregion
export { choice, cidr, domain, email, ensureNonEmptyString, float, hostname, integer, ip, ipv4, ipv6, isNonEmptyString, isValueParser, locale, macAddress, port, portRange, socketAddress, string, url, uuid };
