// @optique/run/dist/run.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const __optique_core_facade = require_rolldown_runtime.__toESM(require("@optique/core/facade"));
const node_path = require_rolldown_runtime.__toESM(require("node:path"));
const node_process = require_rolldown_runtime.__toESM(require("node:process"));

//#region src/run.ts
function run(parserOrProgram, options = {}) {
	return runImpl(parserOrProgram, options);
}
function runSync(parserOrProgram, options = {}) {
	return runImpl(parserOrProgram, options);
}
function runAsync(parserOrProgram, options = {}) {
	const result = runImpl(parserOrProgram, options);
	return Promise.resolve(result);
}
function runImpl(parserOrProgram, options = {}) {
	const isProgram = "parser" in parserOrProgram && "metadata" in parserOrProgram;
	let parser;
	let programNameFromProgram;
	let programMetadata;
	if (isProgram) {
		const program = parserOrProgram;
		parser = program.parser;
		programNameFromProgram = program.metadata.name;
		programMetadata = {
			brief: program.metadata.brief,
			description: program.metadata.description,
			examples: program.metadata.examples,
			author: program.metadata.author,
			bugs: program.metadata.bugs,
			footer: program.metadata.footer
		};
	} else parser = parserOrProgram;
	const { programName = programNameFromProgram ?? node_path.default.basename(node_process.default.argv[1] || "cli"), args = node_process.default.argv.slice(2), colors = node_process.default.stdout.isTTY, maxWidth = node_process.default.stdout.columns, showDefault, showChoices, help, version, completion, aboveError = "usage", errorExitCode = 1, brief = programMetadata?.brief, description = programMetadata?.description, examples = programMetadata?.examples, author = programMetadata?.author, bugs = programMetadata?.bugs, footer = programMetadata?.footer } = options;
	const helpConfig = help ? typeof help === "string" ? {
		mode: help,
		onShow: () => node_process.default.exit(0)
	} : {
		mode: help.mode,
		group: help.group,
		onShow: () => node_process.default.exit(0)
	} : void 0;
	const versionConfig = (() => {
		if (!version) return void 0;
		if (typeof version === "string") return {
			mode: "option",
			value: version,
			onShow: () => node_process.default.exit(0)
		};
		const mode = version.mode ?? "option";
		if (mode === "command" || mode === "both") return {
			mode,
			value: version.value,
			group: version.group,
			onShow: () => node_process.default.exit(0)
		};
		return {
			mode,
			value: version.value,
			onShow: () => node_process.default.exit(0)
		};
	})();
	const completionConfig = (() => {
		if (!completion) return void 0;
		const onShow = () => node_process.default.exit(0);
		if (typeof completion === "string") return {
			mode: completion,
			name: "both",
			helpVisibility: "both",
			onShow
		};
		const mode = completion.mode ?? "both";
		const shells = completion.shells;
		const cGroup = completion.group;
		if (completion.name === "singular") return {
			mode,
			shells,
			...cGroup != null && { group: cGroup },
			name: "singular",
			helpVisibility: completion.helpVisibility ?? "singular",
			onShow
		};
		if (completion.name === "plural") return {
			mode,
			shells,
			...cGroup != null && { group: cGroup },
			name: "plural",
			helpVisibility: completion.helpVisibility ?? "plural",
			onShow
		};
		return {
			mode,
			shells,
			...cGroup != null && { group: cGroup },
			name: "both",
			helpVisibility: completion.helpVisibility ?? "both",
			onShow
		};
	})();
	return (0, __optique_core_facade.runParser)(parser, programName, args, {
		stderr(line) {
			node_process.default.stderr.write(`${line}\n`);
		},
		stdout(line) {
			node_process.default.stdout.write(`${line}\n`);
		},
		colors,
		maxWidth,
		showDefault,
		showChoices,
		help: helpConfig,
		version: versionConfig,
		completion: completionConfig,
		aboveError,
		brief,
		description,
		examples,
		author,
		bugs,
		footer,
		onError() {
			return node_process.default.exit(errorExitCode);
		}
	});
}

//#endregion
exports.run = run;
exports.runAsync = runAsync;
exports.runSync = runSync;
